{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genetic Variant Classifications\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "The ClinVar dataset is a public resource containing annotations about human genetic variants. These variants are (usually manually) classified by clinical laboratories on a categorical spectrum ranging from benign, likely benign, uncertain significance, likely pathogenic, and pathogenic. Variants that have conflicting classifications (from laboratory to laboratory) can cause confusion when clinicians or researchers try to interpret whether the variant has an impact on the disease of a given patient.\n",
    "\n",
    "Conflicting classifications are when two of any of the following three categories are present for one variant, two submissions of one category are not considered conflicting.\n",
    "\n",
    "* Likely Benign or Benign\n",
    "* VUS (uncertain significance)\n",
    "* Likely Pathogenic or Pathogenic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question\n",
    "\n",
    "Given a set of variant features, we are going to try and identify whether that variant is likely to posses a conflicting classification or not using machine learning as opposed to manual classification.\n",
    "\n",
    "Conflicting classification has been assigned to a CLASS column. It is a binary representation of whether or not a variant has conflicting classifications, where 0 represents consistent classifications and 1 represents conflicting classifications."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Models\n",
    "\n",
    "The machine learning models I will be experiementing with to solve this problem are listed below:\n",
    "\n",
    "* Logistic Regression\n",
    "* Random Forest\n",
    "* Gradient Boosting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import ensemble\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from imblearn.ensemble import EasyEnsemble\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.feature_selection import chi2\n",
    "from sklearn.preprocessing import Imputer\n",
    "from sklearn.feature_selection import chi2 as chi2_sk\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "import scipy.stats\n",
    "stats.chisqprob = lambda chisq, df: stats.chi2.sf(chisq, df)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 488,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/rook/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (0,38,40) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(65188, 46)"
      ]
     },
     "execution_count": 488,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('./data/clinvar_conflicting.csv')\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 489,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>...</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>BLOSUM62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>955563</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C3808739,OMIM:615120|MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8|not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.421</td>\n",
       "      <td>11.390</td>\n",
       "      <td>1.133255</td>\n",
       "      <td>-2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>955597</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.42418</td>\n",
       "      <td>0.2826</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.421</td>\n",
       "      <td>8.150</td>\n",
       "      <td>0.599088</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>955619</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.03475</td>\n",
       "      <td>0.0088</td>\n",
       "      <td>MedGen:C3808739,OMIM:615120|MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8|not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.421</td>\n",
       "      <td>3.288</td>\n",
       "      <td>0.069819</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>957640</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0318</td>\n",
       "      <td>0.02016</td>\n",
       "      <td>0.0328</td>\n",
       "      <td>MedGen:C3808739,OMIM:615120|MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8|not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.421</td>\n",
       "      <td>12.560</td>\n",
       "      <td>1.356499</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>976059</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00022</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.421</td>\n",
       "      <td>17.740</td>\n",
       "      <td>2.234711</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM     POS REF ALT  AF_ESP  AF_EXAC  AF_TGP  \\\n",
       "0     1  955563   G   C  0.0000  0.00000  0.0000   \n",
       "1     1  955597   G   T  0.0000  0.42418  0.2826   \n",
       "2     1  955619   G   C  0.0000  0.03475  0.0088   \n",
       "3     1  957640   C   T  0.0318  0.02016  0.0328   \n",
       "4     1  976059   C   T  0.0000  0.00022  0.0010   \n",
       "\n",
       "                                      CLNDISDB CLNDISDBINCL  \\\n",
       "0  MedGen:C3808739,OMIM:615120|MedGen:CN169374          NaN   \n",
       "1                              MedGen:CN169374          NaN   \n",
       "2  MedGen:C3808739,OMIM:615120|MedGen:CN169374          NaN   \n",
       "3  MedGen:C3808739,OMIM:615120|MedGen:CN169374          NaN   \n",
       "4                              MedGen:CN169374          NaN   \n",
       "\n",
       "                                              CLNDN   ...    SIFT PolyPhen  \\\n",
       "0  Myasthenic_syndrome,_congenital,_8|not_specified   ...     NaN      NaN   \n",
       "1                                     not_specified   ...     NaN      NaN   \n",
       "2  Myasthenic_syndrome,_congenital,_8|not_specified   ...     NaN      NaN   \n",
       "3  Myasthenic_syndrome,_congenital,_8|not_specified   ...     NaN      NaN   \n",
       "4                                     not_specified   ...     NaN      NaN   \n",
       "\n",
       "  MOTIF_NAME MOTIF_POS HIGH_INF_POS MOTIF_SCORE_CHANGE  LoFtool  CADD_PHRED  \\\n",
       "0        NaN       NaN          NaN                NaN    0.421      11.390   \n",
       "1        NaN       NaN          NaN                NaN    0.421       8.150   \n",
       "2        NaN       NaN          NaN                NaN    0.421       3.288   \n",
       "3        NaN       NaN          NaN                NaN    0.421      12.560   \n",
       "4        NaN       NaN          NaN                NaN    0.421      17.740   \n",
       "\n",
       "   CADD_RAW BLOSUM62  \n",
       "0  1.133255     -2.0  \n",
       "1  0.599088      NaN  \n",
       "2  0.069819      1.0  \n",
       "3  1.356499      NaN  \n",
       "4  2.234711      NaN  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 489,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Binary class split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    48754\n",
       "1    16434\n",
       "Name: CLASS, dtype: int64"
      ]
     },
     "execution_count": 490,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['CLASS'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 491,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X = df.loc[:, ~df.columns.isin(['CLASS'])]\n",
    "Y = df['CLASS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM\n",
      "25\n",
      "REF\n",
      "866\n",
      "ALT\n",
      "458\n",
      "CLNDISDB\n",
      "9234\n",
      "CLNDISDBINCL\n",
      "48\n",
      "CLNDN\n",
      "9260\n",
      "CLNDNINCL\n",
      "54\n",
      "CLNHGVS\n",
      "65188\n",
      "CLNSIGINCL\n",
      "68\n",
      "CLNVC\n",
      "7\n",
      "CLNVI\n",
      "26289\n",
      "MC\n",
      "89\n",
      "Allele\n",
      "374\n",
      "Consequence\n",
      "48\n",
      "IMPACT\n",
      "4\n",
      "SYMBOL\n",
      "2328\n",
      "Feature_type\n",
      "2\n",
      "Feature\n",
      "2369\n",
      "BIOTYPE\n",
      "2\n",
      "EXON\n",
      "3264\n",
      "INTRON\n",
      "1929\n",
      "cDNA_position\n",
      "13970\n",
      "CDS_position\n",
      "13663\n",
      "Protein_position\n",
      "7339\n",
      "Amino_acids\n",
      "1262\n",
      "Codons\n",
      "2220\n",
      "BAM_EDIT\n",
      "2\n",
      "SIFT\n",
      "4\n",
      "PolyPhen\n",
      "4\n",
      "MOTIF_NAME\n",
      "2\n",
      "HIGH_INF_POS\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "# Converting to categorical\n",
    "convert_cat = []\n",
    "to_drop = []\n",
    "unique = None\n",
    "\n",
    "categorical = X.select_dtypes(include=['object'])\n",
    "for i in categorical:\n",
    "    column = categorical[i]\n",
    "    print(i)\n",
    "    unique = column.nunique()\n",
    "    print(unique)\n",
    "    \n",
    "    cont = pd.crosstab(X[i], Y)\n",
    "    chi2_res = scipy.stats.chi2_contingency(cont)\n",
    "    \n",
    "    # Keep all features with a significant P-value and drop the others\n",
    "    if chi2_res[1] <= 0.05:\n",
    "        convert_cat.append(i)\n",
    "    else:\n",
    "        to_drop.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "convert_cat\n",
      "------\n",
      "['CHROM', 'REF', 'ALT', 'CLNDISDB', 'CLNDN', 'CLNVC', 'MC', 'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Feature', 'EXON', 'INTRON', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons', 'BAM_EDIT'] \n",
      "\n",
      "to_drop\n",
      "------\n",
      "['CLNDISDBINCL', 'CLNDNINCL', 'CLNHGVS', 'CLNSIGINCL', 'CLNVI', 'Feature_type', 'BIOTYPE', 'SIFT', 'PolyPhen', 'MOTIF_NAME', 'HIGH_INF_POS']\n"
     ]
    }
   ],
   "source": [
    "print(f'convert_cat\\n------\\n{convert_cat} \\n')\n",
    "print(f'to_drop\\n------\\n{to_drop}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drop all features with more than 85% NaN's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CLNDISDBINCL: 0.9988341412529913\n",
      "CLNDNINCL: 0.9988341412529913\n",
      "CLNSIGINCL: 0.9988341412529913\n",
      "CLNVI: 0.5757041173222065\n",
      "SSR: 0.9984046143461986\n",
      "INTRON: 0.8649598085537216\n",
      "DISTANCE: 0.998343253359514\n",
      "BAM_EDIT: 0.509587654169479\n",
      "SIFT: 0.6190096336749095\n",
      "PolyPhen: 0.6196232435417561\n",
      "MOTIF_NAME: 0.9999693195066577\n",
      "MOTIF_POS: 0.9999693195066577\n",
      "HIGH_INF_POS: 0.9999693195066577\n",
      "MOTIF_SCORE_CHANGE: 0.9999693195066577\n",
      "BLOSUM62: 0.6073970669448365\n"
     ]
    }
   ],
   "source": [
    "to_drop_nans = []\n",
    "\n",
    "for c in X.columns:\n",
    "    if X[c].isnull().sum() / X.shape[0] > 0.50:\n",
    "        print(f'{c}: {X[c].isnull().sum() / X.shape[0]}')\n",
    "        to_drop_nans.append(c)\n",
    "        \n",
    "        # Remove from the convert_cat array\n",
    "        if c in convert_cat:\n",
    "            convert_cat.remove(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = X.loc[:, ~X.columns.isin(to_drop_nans)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There were a few features that were almost entirely made up of NaN values.  Instead of imputing this data and essentially creating 60,000+ rows of made up figures, I decided to drop them all together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CHROM',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'CLNDISDB',\n",
       " 'CLNDN',\n",
       " 'CLNVC',\n",
       " 'MC',\n",
       " 'Allele',\n",
       " 'Consequence',\n",
       " 'IMPACT',\n",
       " 'SYMBOL',\n",
       " 'Feature',\n",
       " 'EXON',\n",
       " 'cDNA_position',\n",
       " 'CDS_position',\n",
       " 'Protein_position',\n",
       " 'Amino_acids',\n",
       " 'Codons']"
      ]
     },
     "execution_count": 504,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "convert_cat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above are the remaining categorical features to be converted to binary form."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert features to categorical - get_dummies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deleting duplicate rows\n",
    "X = X.loc[:,~X.columns.duplicated()]\n",
    "\n",
    "# Drop cols with too many unique values\n",
    "X = X.loc[:, ~X.columns.isin(to_drop)]\n",
    "\n",
    "# Get dummies - conver to categroical\n",
    "X = pd.get_dummies(data=X, columns=convert_cat)\n",
    "\n",
    "X = X.loc[:,~X.columns.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65188, 66789)"
      ]
     },
     "execution_count": 507,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Find all features created by get_dummies\n",
    "cat_feat = []\n",
    "\n",
    "for feature in convert_cat:\n",
    "    for col in X.loc[:, X.columns.str.startswith(feature + \"_\")].columns:\n",
    "        cat_feat.append(col)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Chi2 after get_dummies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "metadata": {},
   "outputs": [],
   "source": [
    "gd_convert_cat = []\n",
    "gd_to_dropt = []\n",
    "\n",
    "for feature in cat_feat:\n",
    "    cont = pd.crosstab(X[feature], Y)\n",
    "    chi2_res = scipy.stats.chi2_contingency(cont)\n",
    "    \n",
    "    # Keep all features with a significant P-value and drop the others\n",
    "    if chi2_res[1] <= 0.05:\n",
    "        gd_convert_cat.append(feature)\n",
    "    else:\n",
    "        gd_to_dropt.append(feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64912"
      ]
     },
     "execution_count": 545,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gd_to_dropt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the results of the Chi2 test are show that we should drop a substntial number of dummy features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 546,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X = X.loc[:, ~X.columns.isin(gd_to_dropt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65188, 1876)"
      ]
     },
     "execution_count": 547,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above we can see the number of remaining features by examining the shape of X."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding all NaN rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10940"
      ]
     },
     "execution_count": 549,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nans = lambda X: X[X.isnull().any(axis=1)]\n",
    "len(nans(X))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are still quite a few rows with NaN values.  We will have to impute the numerical features."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Numerical Imputer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 553,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['POS', 'AF_ESP', 'AF_EXAC', 'AF_TGP', 'ORIGIN', 'STRAND', 'LoFtool',\n",
      "       'CADD_PHRED', 'CADD_RAW'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "numerical_data = X.loc[:, ~X.columns.isin(X[cat_feat])]\n",
    "print(numerical_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 554,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1d3f6ec358>"
      ]
     },
     "execution_count": 554,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW8AAAEFCAYAAAAsU2YoAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAE5NJREFUeJzt3XmQHGd5x/HvzOzsrna1kiXLJ0YS\nAfJCAga8hATwoSTGVwBTxLmJyziQpGwKClKYmNgECIQk5TgJMQTiFGVIAUnsQKAA26QMVozBRTQY\ngsr2a+SyJQsfuqxjtdccnT96JOvYYzS7s7uv9vspqXanu6f7eeb4qfVOd08hyzIkSWkpzncBkqRj\nZ3hLUoIMb0lKkOEtSQkyvCUpQV1zsZFKpeIhLZLUhsHBwcJE0+ckvJsFtH3fSqUyo/unyJ6Pf4ut\nX7Dndu47GYdNJClBhrckJcjwlqQEGd6SlCDDW5ISZHhLUoIMb0lKkOEtSQkyvCUpQXN2huVMbNg0\nxPbxx46aftGr1855LZK0ELjnLUkJMrwlKUGGtyQlyPCWpAQZ3pKUIMNbkhJkeEtSggxvSUqQ4S1J\nCTK8JSlBhrckJcjwlqQEGd6SlKBpryoYQigDnwXWAnXg7UANuAXIgI3A1THGRseqlCQdppU970uA\nrhjja4APAx8FbgSuizGeAxSASztXoiTpSK2E98NAVwihCCwDqsAgsL45/3bg/M6UJ0maSCtfxjBE\nPmTyELAKeD1wbowxa87fByyfbiWVSqXNEnObt2w+ep3dO2e0zoVupo9ZihZbz4utX7Dn2dJKeL8b\nuDPGeG0I4bnAt4DuQ+YPALunW8ng4GB7FQIbNq1nzeo1E6xzbdvrXOgqlcqMHrMULbaeF1u/YM/t\n3HcyrQybPAPsaf6+CygD94cQ1jWnXQzc01ZlkqS2tLLn/XfAZ0II95Dvcb8f2ADcHELoBh4Ebutc\niZKkI00b3jHGIeA3J5h13uyXI0lqhSfpSFKCDG9JSpDhLUkJMrwlKUGGtyQlyPCWpAQZ3pKUIMNb\nkhJkeEtSggxvSUqQ4S1JCTK8JSlBhrckJcjwlqQEGd6SlCDDW5ISZHhLUoIMb0lKkOEtSQkyvCUp\nQYa3JCXI8JakBBnekpQgw1uSEmR4S1KCDG9JSpDhLUkJMrwlKUGGtyQlyPCWpAQZ3pKUIMNbkhJk\neEtSggxvSUqQ4S1JCTK8JSlBhrckJcjwlqQEGd6SlCDDW5IS1NXKQiGEa4E3At3AJ4H1wC1ABmwE\nro4xNjpUoyTpCNPueYcQ1gGvAV4LnAc8F7gRuC7GeA5QAC7tYI2SpCMUsiybcoEQwsfI97B/HlgG\nvBf4CnBGjDELIVwKXBBjvHqydVQqlak3Mo0Nm4YmnP7KFyydyWolacEbHBwsTDS9lWGTVcAa4PXA\n84CvAsUY44FA3gcsb6GA1iqdwIZN61mzes0E61zb9joXukqlMqPHLEWLrefF1i/Yczv3nUwr4b0T\neCjGOA7EEMIo+dDJAQPA7rYqkyS1pZWjTb4DXBRCKIQQTgf6gbuaY+EAFwP3dKg+SdIEpt3zjjF+\nLYRwLvB98rC/GngUuDmE0A08CNzW0SolSYdp6VDBGOM1E0w+b5ZrkSS1yJN0JClBhrckJcjwlqQE\nGd6SlCDDW5ISZHhLUoIMb0lKkOEtSQkyvCUpQYa3JCXI8JakBBnekpQgw1uSEmR4S1KCDG9JSpDh\nLUkJMrwlKUGGtyQlyPCWpAQZ3pKUIMNbkhJkeEtSggxvSUqQ4S1JCTK8JSlBhrckJcjwlqQEGd6S\nlCDDW5ISZHhLUoIMb0lKkOEtSQkyvCUpQYa3JCXI8JakBBnekpQgw1uSEmR4S1KCDG9JSlBXKwuF\nEE4GKsDrgBpwC5ABG4GrY4yNThUoSTratHveIYQy8GlgpDnpRuC6GOM5QAG4tHPlSZIm0sqwyQ3A\np4AnmrcHgfXN328Hzu9AXZKkKUw5bBJCuALYHmO8M4RwbXNyIcaYNX/fByxvZUOVSqXtIgE2b9l8\n9Dq7d85onQvdTB+zFC22nhdbv2DPs2W6Me8rgSyEcD7wcuBzwMmHzB8AdreyocHBwbYKBNiwaT1r\nVq+ZYJ1r217nQlepVGb0mKVosfW82PoFe27nvpOZctgkxnhujPG8GOM64IfA5cDtIYR1zUUuBu5p\nqypJUttaOtrkCH8C3BxC6AYeBG6b3ZIkSdNpObybe98HnDf7pUiSWuVJOpKUIMNbkhJkeEtSggxv\nSUqQ4S1JCTK8JSlBhrckJcjwlqQEGd6SlCDDW5ISZHhLUoIMb0lKkOEtSQkyvCUpQYa3JCXI8Jak\nBBnekpQgw1uSEmR4S1KCDG9JSpDhLUkJMrwlKUGGtyQlyPCWpAQZ3pKUIMNbkhJkeEtSggxvSUqQ\n4S1JCTK8JSlBhrckJcjwlqQEGd6SlCDDW5ISZHhLUoIMb0lKkOEtSQkyvCUpQYa3JCXI8JakBHVN\nNTOEUAY+A6wFeoCPAA8AtwAZsBG4OsbY6GiVkqTDTLfn/RZgZ4zxHOBi4CbgRuC65rQCcGlnS5Qk\nHWnKPW/gVuC2Q27XgEFgffP27cAFwJen21ClUmmnvoM2b9l89Dq7d85onQvdTB+zFC22nhdbv2DP\ns2XK8I4xDgGEEAbIQ/w64IYYY9ZcZB+wvJUNDQ4Otl3khk3rWbN6zQTrXNv2Ohe6SqUyo8csRYut\n58XWL9hzO/edzLQfWIYQngt8G/jXGOMXgEPHtweA3W1VJUlq25ThHUI4Bfgm8L4Y42eak+8PIaxr\n/n4xcE/nypMkTWS6Me/3AyuA60MI1zenvQv4eAihG3iQw8fEJUlzYLox73eRh/WRzutMOZKkVniS\njiQlyPCWpAQZ3pKUIMNbkhJkeEtSggxvSUqQ4S1JCTK8JSlBhrckJcjwlqQEGd6SlCDDW5ISZHhL\nUoIMb0lKkOEtSQkyvCUpQYa3JCXI8JakBBnekpQgw1uSEmR4S1KCDG9JSpDhLUkJMrwlKUGGtyQl\nyPCWpAQZ3pKUIMNbkhJkeEtSgpIO7588/gwjY7X5LkOS5lyy4b1p627e8/f/wwc+/V3qjWy+y5Gk\nOZVkeA8Nj3N3ZSsAD21+hi99+yfzXJEkza3kwjvLMu7a8Dhj1TqXX/JiVi7r4Qt3PsSjT+yZ79Ik\nac4kF94PPLqLrduGWHvaMi77lRfyzt96BbV6xg2frzA8Wp3v8iRpTiQX3nHLMwCsO+sM7rxvM9uf\nGeGlL1jFlqf28d6P38M37n10niuUpM5LKrxHxmo8tWM/p53YR/+S8sHpZ595OqtPHWDL0/tYf/9W\nsswPMCUd35IK781P7SUD1p62/LDpxWKBC39xDatOWMIDj+7iyr/4Jjfd+kM2PrJjfgqVpA7rmu8C\njsVjT+wFYO3py46a110u8fqzn8f3/u9JntgxxJ33bebO+zZz1otO5ndeFzjlxD76e8t0l0tzXbYk\nzbpkwrteb7Dl6X0sX9rNioGeCZfp7y1z/qtW08gyntyxn/994Gl+8NA2fvDQtoPLnDDQwxknL2Xt\nacs48wWreOnzV7G0r3uu2pCkWdFWeIcQisAngZcBY8DbYoybZrOwI/10+xDVWoO1py2nUChMuWyx\nUOA5Jy3l9HP7efzpIR756W7GqnXGxuvs3T/Oxkd2svGRnXztO/mHm91dRfp6y5ww0MMpK/tYuayX\nWr1BvZHRv6TMioEeVi7rZcVALycM9FAqFqCQ/2OxYqCHYrHA/tEaz+wd5cTlvfT1lqesT5Jmqt09\n7zcBvTHGV4cQfgn4W+DS2Ssrt2vvKJ+87Ufs2DVCo9AA4HkTDJlMplAosPrUAVafOnDY9Fq9wbZd\nw2zdPsTTO4cZHa8xXm3w0+1DPPbk3mOqsVjIh2xGx+vNbcLpq5ayrL+bbc8Ms2dojOVLe1i1fAnd\n5RK1et7Hsv5ulvV3MzRSZeeeERqNjJXLlrC0r8zuoTG2PrmLE76znpNX9LFyeS/lUpFSqUipWKBU\nLFCrZwyPVhkZq1EsFig2p5eKxUN+z/8emJ/XWyD/t69AsZj/bDQyxsZrjI7XKZUKdJdLZI2M4bEa\ntVqDpX15rdVanT1D41TrDVYM9LCsv4f9I1We2TdKlsFAf5n+3jL1Rkat1qCvt4sTBnroLpfYt3+c\n/SNVurpK9PV2Ua832D00zvBolZ5yiSW9XWzePMTjQ5toNDLKXSW6yyW6y0W6u0qUy0W6u/LHoFpt\nMFatMVZtMF6t02hkdJWKlLuKdHU1fzZvFwsFxsbrjIzXIINiERoZjFfrjFcb9JSL9PZ0kWUZI2N1\navUGS3q66OvtIsugWquTZdDVXNfoeI3h0fySDD3lEr3dJXq681rrjaxZW51qrU611qBQyB/zA89L\nsZjfLhYLPPLEKMWl28gy8tdgrUFXqUC59GwfxUKBaq1BtdZgvLnOLMu3feCxOfC6GhmrUa036CmX\n6CmXaGRZvhNSz6jWG9TrDWr1BrV6Rm93ieX9PfQtyftsNDIaWXbwZ5a/TOnpKbGku4tGluXrrzUO\nvo7I/xzcmconNV9fk8x7ctc4D23exchojbFq/eBjUSwWKDXXmWUZWZa/r/p6uygWC0fVeGCZevP3\nAzXXs4xyqciSni7K5SKjY/nzlTWf+2KxcHCbB14jAGPVOvV6Rrn5uNfqDUbH6gffq329ZUbGquwb\nrlIoQF9vmWIBdu4ZZdfeUfqXlDllZR/FQoGt24fYuXuEk1Ys4bmnDNDo0AEU7Yb32cAdADHG+0II\nr5y9kp41PFrlx4/saL5ZxugplzjtxP4Zr7erVOT0k5Zy+klLD5ueZRmj43WGR6sH32xj4zX2j9YY\nHq0yPFpjeKyWH82Sweh4nf2jVaq1Oqes7KOvt8ze/eNs3z3MEzuGWLqkzMplvYyM1fnJ47tpZBnF\nAjTvflCxWKAAbNq655AaYce+PTy8ZfeM+03LIuv37kX4ofod26Zf5jjyiuf38QsdSMh2w3sZcOgp\njfUQQleMcdKrRFUqlbY2dM2bTz1iyq621tOyHmBgspldTP+Q9QEnzGZFkhLXbv5Npd3w3svhEVec\nKrgHBwenHqSWJB2Tdo/zvhe4BKA55v3jWatIkjStdve8vwy8LoTwXfLPJd46eyVJkqZT8FRySUpP\nUqfHS5JyhrckJcjwlqQELZhrm0x3yn0I4e3AHwE14CMxxq/NS6GzqJXLDIQQTgK+C7w0xjg691XO\nrhae53cDv928+Y0Y44fmvsrZ1ULPVwNXkJ+79eFF9NouAl8HvhJj/NTcVzm7WniePw68FtjXnHRp\njLHtrwBbSHveB0+5B/6U/JR7AEIIpwLvJG/8QuBjIYSJr06Vlkl7BgghXAh8EzhlHmrrlKme558B\nfg94DfBq4IIQwpnzUuXsmqrnVcBV5D3/KvBPIYTj4byIKV/bTR8BVs5pVZ01Xc9nARfGGNc1/87o\nuxsXUngfdso9cOgJpa8C7o0xjjUb3gQcD2/qqXoGaADn0/HTSufUVD0/DlwUY6zHGBtAGUj+fxtM\n0XOMcQfwshhjFTgV2B1jPB4OAZvytR1CuIz89X373JfWMZP23NwrfyHwzyGEe0MIV850YwspvCc8\n5X6SefuAw7+RIU1T9UyM8b9jjDvnvqyOmrTnGGM1xrgjhFAIIdwA3B9jfHheqpxd0z3PtRDCO4D7\ngNvmurgOmbTnEMJLgN8FPjAfhXXQVM9zP/CPwFuAi4CrZvq/yoUU3lOdcn/kvAGOjysYHdNlBo4T\nU/YcQugFPt9c5qo5rq1Tpn2eY4w3AacB54YQfnkui+uQqXq+HHgO8C3ysf73hBAumtvyOmKqnoeB\nf4gxDscY95H3/rKZbGzBfGBJfsr9G4D/mOCU++8DH22+sXuAFwMb577EWTdVz8erSXtujvV+BfhW\njPGv56m+Tpiq5wB8DPh1oEr+QVdjPoqcZZP2HGO85sDvIYQPAk/FGO+Y8wpn31Tv558F/i2EcBb5\nTvPZwGdnsrEFc4blIZ/Unsmzp9xfAmyKMX61ebTJH5I3/pcxxv+ct2JnyXQ9H7LcY8CLjrOjTY7q\nGSgBXyQfPjjg2hjj9+a6ztnUwmv7z4GLyY82uT3G+OF5K3aWHMNr+4Pk4X08HW0y2fN8DfAb5P9I\nf26mPS+Y8JYktW4hjXlLklpkeEtSggxvSUqQ4S1JCTK8JSlBhrckJWghnaQjHZPmadY/Bi47cNx/\nCOFu4Axg6JBFb44xfmKK9WTAj46Y/PUY45+FEFYDnwDWkO/sPAC8I8a4LYRwBXAjsKV5nyXAeuCq\nRXCmrOaZ4a2UXQncSn6p4ENP2npbjPHuY1lRjPHlk8z6NPkJFV8ECCFcC3wKeHNz/ldjjFc055XI\nL9/7B837SR3jsImSFEIok18+9jrgrBDC8zu0qVOBvkNu39T8e5QYYx24B3hJh2qRDnLPW6n6NWBz\njPHhEMJ/kV864X3Nef8SQjgwbLIvxnjOdCsLIfzwiEnvizHeCVwLfD6E8CHgLvJLmN46yTpOBC4A\n/uqYu5GOkeGtVL2V/DooAP9OHrDXN2/P2rBJjPGOEMJzgHXk11b/G/Jv+nlTc5E3NoO/QP4/2S8d\nUpfUMYa3khNCOJn8Qk6DIYR3kQfnCp4dh56t7awEro8xvpv8Ivt3hBD+Aniy+fV0cMiYtzSXHPNW\nin4fuCvGeEaMcW2McQ3wUeCPZ3k7e8j3rC8/ZNrPAU9zfH27kRLknrdSdAXw/iOmfQK4hvyC+Mds\ngjHvTTHGy0IIlwA3Nve4h4EngDfEGOv5pbil+eElYSUpQe5567jXPIxwsi/veFuMccNc1iPNBve8\nJSlBfmApSQkyvCUpQYa3JCXI8JakBP0/0btOrBAtDegAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1d2bea9e10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(X['AF_ESP'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that these numerical features are not normally distributed.  So we will use the median to impute the data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 557,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For numerical data, impute using mean OR median values\n",
    "imp = Imputer(missing_values='NaN', strategy='median', axis=0)\n",
    "imp = imp.fit(numerical_data)\n",
    "\n",
    "# Impute our data\n",
    "X[numerical_data.columns] = imp.transform(numerical_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 558,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 558,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nans = lambda X: X[X.isnull().any(axis=1)]\n",
    "len(nans(X))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that we no longer have any NaN values in our data.\n",
    "\n",
    "We do not need to impute the categorical features because get_dummies does this for us."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VarianceThreshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 562,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>955563.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.390</td>\n",
       "      <td>1.133255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>955597.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.150</td>\n",
       "      <td>0.599088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>955619.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.288</td>\n",
       "      <td>0.069819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>957640.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.560</td>\n",
       "      <td>1.356499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>976059.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17.740</td>\n",
       "      <td>2.234711</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        POS  ORIGIN  STRAND  CADD_PHRED  CADD_RAW\n",
       "0  955563.0     1.0     1.0      11.390  1.133255\n",
       "1  955597.0     1.0     1.0       8.150  0.599088\n",
       "2  955619.0     1.0     1.0       3.288  0.069819\n",
       "3  957640.0     1.0     1.0      12.560  1.356499\n",
       "4  976059.0     1.0     1.0      17.740  2.234711"
      ]
     },
     "execution_count": 562,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Removes all low-variance features\n",
    "def variance_threshold_selector(data, threshold=0.95):\n",
    "    selector = VarianceThreshold(threshold)\n",
    "    selector.fit(data)\n",
    "    return data[data.columns[selector.get_support(indices=True)]]\n",
    "\n",
    "vt_to_drop = variance_threshold_selector(X)\n",
    "\n",
    "vt_to_drop.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 563,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X = X.loc[:, ~X.columns.isin(vt_to_drop)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Variables with low variance really aren’t beneficial to our model."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Display correlation Matrix to identify features that need to be dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 564,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CHROM_1</th>\n",
       "      <th>CHROM_2</th>\n",
       "      <th>CHROM_5</th>\n",
       "      <th>CHROM_7</th>\n",
       "      <th>CHROM_10</th>\n",
       "      <th>CHROM_11</th>\n",
       "      <th>...</th>\n",
       "      <th>Codons_tCt/tGt</th>\n",
       "      <th>Codons_tGg/tAg</th>\n",
       "      <th>Codons_tTa/tGa</th>\n",
       "      <th>Codons_taC/taA</th>\n",
       "      <th>Codons_taC/taG</th>\n",
       "      <th>Codons_taT/taA</th>\n",
       "      <th>Codons_tcG/tcA</th>\n",
       "      <th>Codons_tgC/tgA</th>\n",
       "      <th>Codons_tgG/tgA</th>\n",
       "      <th>Codons_ttC/ttT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AF_ESP</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.851870</td>\n",
       "      <td>0.807741</td>\n",
       "      <td>0.024530</td>\n",
       "      <td>0.023263</td>\n",
       "      <td>-0.021822</td>\n",
       "      <td>0.006211</td>\n",
       "      <td>0.001599</td>\n",
       "      <td>0.011395</td>\n",
       "      <td>-0.022898</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007952</td>\n",
       "      <td>-0.008781</td>\n",
       "      <td>-0.004284</td>\n",
       "      <td>-0.007166</td>\n",
       "      <td>-0.006793</td>\n",
       "      <td>-0.004711</td>\n",
       "      <td>0.004803</td>\n",
       "      <td>-0.006221</td>\n",
       "      <td>-0.008826</td>\n",
       "      <td>0.009074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_EXAC</th>\n",
       "      <td>0.851870</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.805634</td>\n",
       "      <td>0.023114</td>\n",
       "      <td>0.016571</td>\n",
       "      <td>-0.018818</td>\n",
       "      <td>-0.001123</td>\n",
       "      <td>0.003508</td>\n",
       "      <td>0.014726</td>\n",
       "      <td>-0.020768</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007627</td>\n",
       "      <td>-0.008528</td>\n",
       "      <td>-0.004155</td>\n",
       "      <td>-0.006839</td>\n",
       "      <td>-0.006537</td>\n",
       "      <td>-0.004571</td>\n",
       "      <td>0.005235</td>\n",
       "      <td>-0.006065</td>\n",
       "      <td>-0.008504</td>\n",
       "      <td>0.005833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_TGP</th>\n",
       "      <td>0.807741</td>\n",
       "      <td>0.805634</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.025189</td>\n",
       "      <td>0.018733</td>\n",
       "      <td>-0.023872</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.001747</td>\n",
       "      <td>0.019362</td>\n",
       "      <td>-0.022115</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.008289</td>\n",
       "      <td>-0.008683</td>\n",
       "      <td>-0.004378</td>\n",
       "      <td>-0.007121</td>\n",
       "      <td>-0.006902</td>\n",
       "      <td>-0.004817</td>\n",
       "      <td>0.005580</td>\n",
       "      <td>-0.006394</td>\n",
       "      <td>-0.008986</td>\n",
       "      <td>0.006200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LoFtool</th>\n",
       "      <td>0.024530</td>\n",
       "      <td>0.023114</td>\n",
       "      <td>0.025189</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.042282</td>\n",
       "      <td>0.186457</td>\n",
       "      <td>0.015560</td>\n",
       "      <td>-0.041412</td>\n",
       "      <td>-0.009614</td>\n",
       "      <td>0.121785</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002127</td>\n",
       "      <td>-0.002202</td>\n",
       "      <td>0.005689</td>\n",
       "      <td>-0.000901</td>\n",
       "      <td>0.000648</td>\n",
       "      <td>0.001360</td>\n",
       "      <td>-0.002478</td>\n",
       "      <td>-0.010072</td>\n",
       "      <td>-0.002022</td>\n",
       "      <td>-0.004718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_1</th>\n",
       "      <td>0.023263</td>\n",
       "      <td>0.016571</td>\n",
       "      <td>0.018733</td>\n",
       "      <td>-0.042282</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.105889</td>\n",
       "      <td>-0.065886</td>\n",
       "      <td>-0.058084</td>\n",
       "      <td>-0.052266</td>\n",
       "      <td>-0.076743</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004973</td>\n",
       "      <td>0.002776</td>\n",
       "      <td>-0.001062</td>\n",
       "      <td>-0.005684</td>\n",
       "      <td>-0.002868</td>\n",
       "      <td>0.001387</td>\n",
       "      <td>-0.004161</td>\n",
       "      <td>-0.004369</td>\n",
       "      <td>0.000804</td>\n",
       "      <td>0.004612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_2</th>\n",
       "      <td>-0.021822</td>\n",
       "      <td>-0.018818</td>\n",
       "      <td>-0.023872</td>\n",
       "      <td>0.186457</td>\n",
       "      <td>-0.105889</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.095132</td>\n",
       "      <td>-0.083867</td>\n",
       "      <td>-0.075466</td>\n",
       "      <td>-0.110809</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007609</td>\n",
       "      <td>-0.003856</td>\n",
       "      <td>-0.001377</td>\n",
       "      <td>-0.001826</td>\n",
       "      <td>0.001058</td>\n",
       "      <td>0.004696</td>\n",
       "      <td>-0.010003</td>\n",
       "      <td>-0.002593</td>\n",
       "      <td>-0.006088</td>\n",
       "      <td>-0.002884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_5</th>\n",
       "      <td>0.006211</td>\n",
       "      <td>-0.001123</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.015560</td>\n",
       "      <td>-0.065886</td>\n",
       "      <td>-0.095132</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.052183</td>\n",
       "      <td>-0.046956</td>\n",
       "      <td>-0.068947</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014881</td>\n",
       "      <td>-0.002030</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>-0.002363</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>-0.001015</td>\n",
       "      <td>-0.004765</td>\n",
       "      <td>-0.003440</td>\n",
       "      <td>-0.006686</td>\n",
       "      <td>-0.006659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_7</th>\n",
       "      <td>0.001599</td>\n",
       "      <td>0.003508</td>\n",
       "      <td>0.001747</td>\n",
       "      <td>-0.041412</td>\n",
       "      <td>-0.058084</td>\n",
       "      <td>-0.083867</td>\n",
       "      <td>-0.052183</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.041396</td>\n",
       "      <td>-0.060782</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000543</td>\n",
       "      <td>-0.002085</td>\n",
       "      <td>0.000720</td>\n",
       "      <td>-0.003575</td>\n",
       "      <td>-0.003064</td>\n",
       "      <td>-0.004030</td>\n",
       "      <td>0.007930</td>\n",
       "      <td>-0.005381</td>\n",
       "      <td>-0.005442</td>\n",
       "      <td>0.007095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_10</th>\n",
       "      <td>0.011395</td>\n",
       "      <td>0.014726</td>\n",
       "      <td>0.019362</td>\n",
       "      <td>-0.009614</td>\n",
       "      <td>-0.052266</td>\n",
       "      <td>-0.075466</td>\n",
       "      <td>-0.046956</td>\n",
       "      <td>-0.041396</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.054694</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007826</td>\n",
       "      <td>-0.005060</td>\n",
       "      <td>-0.003295</td>\n",
       "      <td>-0.002691</td>\n",
       "      <td>-0.002200</td>\n",
       "      <td>0.000764</td>\n",
       "      <td>0.003458</td>\n",
       "      <td>-0.001553</td>\n",
       "      <td>-0.004467</td>\n",
       "      <td>-0.000402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_11</th>\n",
       "      <td>-0.022898</td>\n",
       "      <td>-0.020768</td>\n",
       "      <td>-0.022115</td>\n",
       "      <td>0.121785</td>\n",
       "      <td>-0.076743</td>\n",
       "      <td>-0.110809</td>\n",
       "      <td>-0.068947</td>\n",
       "      <td>-0.060782</td>\n",
       "      <td>-0.054694</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001511</td>\n",
       "      <td>0.015828</td>\n",
       "      <td>0.012290</td>\n",
       "      <td>0.004036</td>\n",
       "      <td>0.009554</td>\n",
       "      <td>-0.002210</td>\n",
       "      <td>-0.005767</td>\n",
       "      <td>0.002221</td>\n",
       "      <td>0.011584</td>\n",
       "      <td>-0.001269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_13</th>\n",
       "      <td>-0.027537</td>\n",
       "      <td>-0.027953</td>\n",
       "      <td>-0.028789</td>\n",
       "      <td>-0.119747</td>\n",
       "      <td>-0.055702</td>\n",
       "      <td>-0.080427</td>\n",
       "      <td>-0.050043</td>\n",
       "      <td>-0.044117</td>\n",
       "      <td>-0.039698</td>\n",
       "      <td>-0.058290</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010863</td>\n",
       "      <td>0.004692</td>\n",
       "      <td>0.005596</td>\n",
       "      <td>-0.003220</td>\n",
       "      <td>-0.005583</td>\n",
       "      <td>0.000275</td>\n",
       "      <td>-0.014818</td>\n",
       "      <td>-0.002059</td>\n",
       "      <td>-0.000635</td>\n",
       "      <td>-0.013908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_14</th>\n",
       "      <td>0.009595</td>\n",
       "      <td>0.005834</td>\n",
       "      <td>0.011515</td>\n",
       "      <td>-0.029261</td>\n",
       "      <td>-0.047855</td>\n",
       "      <td>-0.069097</td>\n",
       "      <td>-0.042993</td>\n",
       "      <td>-0.037902</td>\n",
       "      <td>-0.034106</td>\n",
       "      <td>-0.050078</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.002742</td>\n",
       "      <td>0.002895</td>\n",
       "      <td>-0.003017</td>\n",
       "      <td>0.004247</td>\n",
       "      <td>-0.004797</td>\n",
       "      <td>0.001447</td>\n",
       "      <td>0.015643</td>\n",
       "      <td>-0.004433</td>\n",
       "      <td>-0.006233</td>\n",
       "      <td>-0.007929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_15</th>\n",
       "      <td>0.004683</td>\n",
       "      <td>0.003479</td>\n",
       "      <td>0.007204</td>\n",
       "      <td>-0.024928</td>\n",
       "      <td>-0.044647</td>\n",
       "      <td>-0.064465</td>\n",
       "      <td>-0.040111</td>\n",
       "      <td>-0.035362</td>\n",
       "      <td>-0.031819</td>\n",
       "      <td>-0.046721</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>-0.006198</td>\n",
       "      <td>0.002784</td>\n",
       "      <td>-0.004747</td>\n",
       "      <td>0.002572</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>0.001047</td>\n",
       "      <td>0.003488</td>\n",
       "      <td>-0.003102</td>\n",
       "      <td>-0.004093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_16</th>\n",
       "      <td>-0.009229</td>\n",
       "      <td>-0.013755</td>\n",
       "      <td>-0.012652</td>\n",
       "      <td>0.042957</td>\n",
       "      <td>-0.057059</td>\n",
       "      <td>-0.082387</td>\n",
       "      <td>-0.051263</td>\n",
       "      <td>-0.045192</td>\n",
       "      <td>-0.040665</td>\n",
       "      <td>-0.059710</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000848</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000857</td>\n",
       "      <td>-0.003424</td>\n",
       "      <td>-0.000113</td>\n",
       "      <td>0.004139</td>\n",
       "      <td>0.012596</td>\n",
       "      <td>-0.005286</td>\n",
       "      <td>-0.003115</td>\n",
       "      <td>0.004594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_17</th>\n",
       "      <td>-0.030896</td>\n",
       "      <td>-0.031693</td>\n",
       "      <td>-0.026714</td>\n",
       "      <td>-0.091613</td>\n",
       "      <td>-0.081337</td>\n",
       "      <td>-0.117441</td>\n",
       "      <td>-0.073073</td>\n",
       "      <td>-0.064420</td>\n",
       "      <td>-0.057968</td>\n",
       "      <td>-0.085115</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011205</td>\n",
       "      <td>-0.000909</td>\n",
       "      <td>0.001396</td>\n",
       "      <td>0.001029</td>\n",
       "      <td>0.004163</td>\n",
       "      <td>-0.005643</td>\n",
       "      <td>-0.010023</td>\n",
       "      <td>0.001349</td>\n",
       "      <td>0.005212</td>\n",
       "      <td>-0.001894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_20</th>\n",
       "      <td>0.016523</td>\n",
       "      <td>0.022718</td>\n",
       "      <td>0.018930</td>\n",
       "      <td>-0.040403</td>\n",
       "      <td>-0.029393</td>\n",
       "      <td>-0.042440</td>\n",
       "      <td>-0.026407</td>\n",
       "      <td>-0.023280</td>\n",
       "      <td>-0.020948</td>\n",
       "      <td>-0.030758</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004401</td>\n",
       "      <td>-0.004080</td>\n",
       "      <td>-0.001853</td>\n",
       "      <td>-0.003125</td>\n",
       "      <td>0.002326</td>\n",
       "      <td>-0.002039</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>0.002981</td>\n",
       "      <td>-0.003828</td>\n",
       "      <td>0.013806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_21</th>\n",
       "      <td>0.033424</td>\n",
       "      <td>0.026051</td>\n",
       "      <td>0.027190</td>\n",
       "      <td>-0.033917</td>\n",
       "      <td>-0.030997</td>\n",
       "      <td>-0.044756</td>\n",
       "      <td>-0.027848</td>\n",
       "      <td>-0.024550</td>\n",
       "      <td>-0.022091</td>\n",
       "      <td>-0.032437</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004641</td>\n",
       "      <td>-0.004303</td>\n",
       "      <td>-0.001954</td>\n",
       "      <td>-0.003296</td>\n",
       "      <td>-0.003107</td>\n",
       "      <td>-0.002150</td>\n",
       "      <td>0.001835</td>\n",
       "      <td>-0.002871</td>\n",
       "      <td>-0.000183</td>\n",
       "      <td>-0.004589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_22</th>\n",
       "      <td>0.003739</td>\n",
       "      <td>0.002169</td>\n",
       "      <td>0.005391</td>\n",
       "      <td>-0.011221</td>\n",
       "      <td>-0.036049</td>\n",
       "      <td>-0.052050</td>\n",
       "      <td>-0.032386</td>\n",
       "      <td>-0.028551</td>\n",
       "      <td>-0.025691</td>\n",
       "      <td>-0.037723</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000397</td>\n",
       "      <td>-0.001880</td>\n",
       "      <td>-0.002273</td>\n",
       "      <td>0.004320</td>\n",
       "      <td>-0.003613</td>\n",
       "      <td>-0.002501</td>\n",
       "      <td>0.002014</td>\n",
       "      <td>0.001339</td>\n",
       "      <td>0.001963</td>\n",
       "      <td>0.003255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM_MT</th>\n",
       "      <td>-0.003934</td>\n",
       "      <td>-0.003814</td>\n",
       "      <td>-0.004018</td>\n",
       "      <td>-0.007821</td>\n",
       "      <td>-0.004243</td>\n",
       "      <td>-0.006127</td>\n",
       "      <td>-0.003812</td>\n",
       "      <td>-0.003361</td>\n",
       "      <td>-0.003024</td>\n",
       "      <td>-0.004440</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000635</td>\n",
       "      <td>-0.000589</td>\n",
       "      <td>-0.000268</td>\n",
       "      <td>-0.000451</td>\n",
       "      <td>-0.000425</td>\n",
       "      <td>-0.000294</td>\n",
       "      <td>-0.001206</td>\n",
       "      <td>-0.000393</td>\n",
       "      <td>-0.000553</td>\n",
       "      <td>-0.001141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_A</th>\n",
       "      <td>0.018794</td>\n",
       "      <td>0.016609</td>\n",
       "      <td>0.012548</td>\n",
       "      <td>0.022992</td>\n",
       "      <td>-0.018791</td>\n",
       "      <td>0.009397</td>\n",
       "      <td>0.047899</td>\n",
       "      <td>-0.001251</td>\n",
       "      <td>0.001485</td>\n",
       "      <td>0.019456</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.017102</td>\n",
       "      <td>-0.015856</td>\n",
       "      <td>0.012876</td>\n",
       "      <td>-0.012144</td>\n",
       "      <td>-0.011449</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>-0.032467</td>\n",
       "      <td>-0.010581</td>\n",
       "      <td>-0.014877</td>\n",
       "      <td>-0.030720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_AC</th>\n",
       "      <td>-0.007512</td>\n",
       "      <td>-0.005643</td>\n",
       "      <td>-0.005694</td>\n",
       "      <td>-0.004787</td>\n",
       "      <td>-0.007711</td>\n",
       "      <td>-0.004167</td>\n",
       "      <td>-0.004634</td>\n",
       "      <td>-0.004765</td>\n",
       "      <td>-0.007678</td>\n",
       "      <td>-0.002440</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001613</td>\n",
       "      <td>-0.001496</td>\n",
       "      <td>-0.000679</td>\n",
       "      <td>-0.001145</td>\n",
       "      <td>-0.001080</td>\n",
       "      <td>-0.000747</td>\n",
       "      <td>-0.003062</td>\n",
       "      <td>-0.000998</td>\n",
       "      <td>-0.001403</td>\n",
       "      <td>-0.002897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_C</th>\n",
       "      <td>-0.010447</td>\n",
       "      <td>-0.011472</td>\n",
       "      <td>-0.011046</td>\n",
       "      <td>-0.022567</td>\n",
       "      <td>0.012841</td>\n",
       "      <td>-0.003046</td>\n",
       "      <td>-0.018144</td>\n",
       "      <td>-0.007245</td>\n",
       "      <td>-0.001712</td>\n",
       "      <td>-0.011086</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017045</td>\n",
       "      <td>0.016662</td>\n",
       "      <td>-0.012102</td>\n",
       "      <td>0.006717</td>\n",
       "      <td>0.008330</td>\n",
       "      <td>-0.013316</td>\n",
       "      <td>0.025323</td>\n",
       "      <td>0.013346</td>\n",
       "      <td>0.013766</td>\n",
       "      <td>0.027364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_CAAA</th>\n",
       "      <td>-0.001703</td>\n",
       "      <td>-0.001630</td>\n",
       "      <td>-0.001694</td>\n",
       "      <td>0.005030</td>\n",
       "      <td>-0.001837</td>\n",
       "      <td>-0.002653</td>\n",
       "      <td>0.008194</td>\n",
       "      <td>0.009573</td>\n",
       "      <td>-0.001309</td>\n",
       "      <td>-0.001923</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000275</td>\n",
       "      <td>-0.000255</td>\n",
       "      <td>-0.000116</td>\n",
       "      <td>-0.000195</td>\n",
       "      <td>-0.000184</td>\n",
       "      <td>-0.000127</td>\n",
       "      <td>-0.000522</td>\n",
       "      <td>-0.000170</td>\n",
       "      <td>-0.000239</td>\n",
       "      <td>-0.000494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_CAG</th>\n",
       "      <td>-0.005832</td>\n",
       "      <td>-0.005272</td>\n",
       "      <td>-0.005566</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.003295</td>\n",
       "      <td>-0.006088</td>\n",
       "      <td>0.000358</td>\n",
       "      <td>-0.000650</td>\n",
       "      <td>-0.005400</td>\n",
       "      <td>0.012986</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001135</td>\n",
       "      <td>-0.001052</td>\n",
       "      <td>-0.000478</td>\n",
       "      <td>-0.000806</td>\n",
       "      <td>-0.000760</td>\n",
       "      <td>-0.000526</td>\n",
       "      <td>-0.002154</td>\n",
       "      <td>-0.000702</td>\n",
       "      <td>-0.000987</td>\n",
       "      <td>-0.002038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_CG</th>\n",
       "      <td>-0.007103</td>\n",
       "      <td>-0.006876</td>\n",
       "      <td>-0.007484</td>\n",
       "      <td>-0.011411</td>\n",
       "      <td>-0.001964</td>\n",
       "      <td>-0.007117</td>\n",
       "      <td>0.006179</td>\n",
       "      <td>0.001127</td>\n",
       "      <td>-0.000229</td>\n",
       "      <td>0.009196</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001210</td>\n",
       "      <td>-0.001122</td>\n",
       "      <td>-0.000510</td>\n",
       "      <td>-0.000859</td>\n",
       "      <td>-0.000810</td>\n",
       "      <td>-0.000561</td>\n",
       "      <td>-0.002297</td>\n",
       "      <td>-0.000749</td>\n",
       "      <td>-0.001053</td>\n",
       "      <td>-0.002174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_CT</th>\n",
       "      <td>-0.005472</td>\n",
       "      <td>-0.006068</td>\n",
       "      <td>-0.008346</td>\n",
       "      <td>0.001048</td>\n",
       "      <td>-0.006380</td>\n",
       "      <td>-0.008969</td>\n",
       "      <td>-0.004625</td>\n",
       "      <td>0.007594</td>\n",
       "      <td>-0.002862</td>\n",
       "      <td>0.010163</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001784</td>\n",
       "      <td>-0.001654</td>\n",
       "      <td>-0.000751</td>\n",
       "      <td>-0.001267</td>\n",
       "      <td>-0.001195</td>\n",
       "      <td>-0.000827</td>\n",
       "      <td>-0.003388</td>\n",
       "      <td>-0.001104</td>\n",
       "      <td>-0.001552</td>\n",
       "      <td>-0.003205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_CTT</th>\n",
       "      <td>-0.002702</td>\n",
       "      <td>-0.000383</td>\n",
       "      <td>-0.001014</td>\n",
       "      <td>0.008704</td>\n",
       "      <td>0.002694</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>-0.002107</td>\n",
       "      <td>-0.004602</td>\n",
       "      <td>-0.004141</td>\n",
       "      <td>0.015732</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000870</td>\n",
       "      <td>-0.000807</td>\n",
       "      <td>-0.000366</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>-0.000582</td>\n",
       "      <td>-0.000403</td>\n",
       "      <td>-0.001652</td>\n",
       "      <td>-0.000538</td>\n",
       "      <td>-0.000757</td>\n",
       "      <td>-0.001563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_G</th>\n",
       "      <td>-0.012746</td>\n",
       "      <td>-0.010841</td>\n",
       "      <td>-0.010406</td>\n",
       "      <td>-0.032604</td>\n",
       "      <td>0.006412</td>\n",
       "      <td>-0.013088</td>\n",
       "      <td>-0.025718</td>\n",
       "      <td>-0.001668</td>\n",
       "      <td>0.008588</td>\n",
       "      <td>-0.011207</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012060</td>\n",
       "      <td>0.010319</td>\n",
       "      <td>-0.011921</td>\n",
       "      <td>0.013979</td>\n",
       "      <td>0.011171</td>\n",
       "      <td>-0.013116</td>\n",
       "      <td>0.029969</td>\n",
       "      <td>0.004647</td>\n",
       "      <td>0.011558</td>\n",
       "      <td>0.024935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_GA</th>\n",
       "      <td>-0.001044</td>\n",
       "      <td>0.000436</td>\n",
       "      <td>0.003071</td>\n",
       "      <td>-0.001055</td>\n",
       "      <td>-0.003611</td>\n",
       "      <td>0.004763</td>\n",
       "      <td>-0.003730</td>\n",
       "      <td>-0.000004</td>\n",
       "      <td>-0.002800</td>\n",
       "      <td>-0.002764</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001516</td>\n",
       "      <td>-0.001406</td>\n",
       "      <td>-0.000638</td>\n",
       "      <td>-0.001077</td>\n",
       "      <td>-0.001015</td>\n",
       "      <td>-0.000702</td>\n",
       "      <td>-0.002878</td>\n",
       "      <td>-0.000938</td>\n",
       "      <td>-0.001319</td>\n",
       "      <td>-0.002723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REF_GC</th>\n",
       "      <td>-0.009958</td>\n",
       "      <td>-0.009661</td>\n",
       "      <td>-0.008387</td>\n",
       "      <td>-0.007193</td>\n",
       "      <td>-0.005438</td>\n",
       "      <td>-0.004334</td>\n",
       "      <td>0.001100</td>\n",
       "      <td>-0.001744</td>\n",
       "      <td>-0.004079</td>\n",
       "      <td>0.006466</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001690</td>\n",
       "      <td>-0.001567</td>\n",
       "      <td>-0.000712</td>\n",
       "      <td>-0.001200</td>\n",
       "      <td>-0.001131</td>\n",
       "      <td>-0.000783</td>\n",
       "      <td>-0.003208</td>\n",
       "      <td>-0.001045</td>\n",
       "      <td>-0.001470</td>\n",
       "      <td>-0.003035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_agC/agA</th>\n",
       "      <td>-0.003750</td>\n",
       "      <td>-0.003521</td>\n",
       "      <td>-0.004138</td>\n",
       "      <td>-0.002935</td>\n",
       "      <td>-0.003778</td>\n",
       "      <td>0.000435</td>\n",
       "      <td>-0.002876</td>\n",
       "      <td>0.004511</td>\n",
       "      <td>-0.004537</td>\n",
       "      <td>0.000806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000953</td>\n",
       "      <td>-0.000884</td>\n",
       "      <td>-0.000401</td>\n",
       "      <td>-0.000677</td>\n",
       "      <td>-0.000638</td>\n",
       "      <td>-0.000442</td>\n",
       "      <td>-0.001809</td>\n",
       "      <td>-0.000590</td>\n",
       "      <td>-0.000829</td>\n",
       "      <td>-0.001712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_agT/agC</th>\n",
       "      <td>0.014429</td>\n",
       "      <td>0.013989</td>\n",
       "      <td>0.008679</td>\n",
       "      <td>0.007449</td>\n",
       "      <td>0.003616</td>\n",
       "      <td>-0.003498</td>\n",
       "      <td>0.012968</td>\n",
       "      <td>-0.000599</td>\n",
       "      <td>-0.001105</td>\n",
       "      <td>0.005607</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001573</td>\n",
       "      <td>-0.001459</td>\n",
       "      <td>-0.000663</td>\n",
       "      <td>-0.001117</td>\n",
       "      <td>-0.001053</td>\n",
       "      <td>-0.000729</td>\n",
       "      <td>-0.002987</td>\n",
       "      <td>-0.000973</td>\n",
       "      <td>-0.001369</td>\n",
       "      <td>-0.002826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_atC/atT</th>\n",
       "      <td>0.000906</td>\n",
       "      <td>0.000755</td>\n",
       "      <td>0.005853</td>\n",
       "      <td>-0.012172</td>\n",
       "      <td>0.005836</td>\n",
       "      <td>-0.000267</td>\n",
       "      <td>-0.011117</td>\n",
       "      <td>-0.003099</td>\n",
       "      <td>0.007204</td>\n",
       "      <td>-0.009120</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003346</td>\n",
       "      <td>-0.003103</td>\n",
       "      <td>-0.001409</td>\n",
       "      <td>-0.002376</td>\n",
       "      <td>-0.002240</td>\n",
       "      <td>-0.001550</td>\n",
       "      <td>-0.006353</td>\n",
       "      <td>-0.002070</td>\n",
       "      <td>-0.002911</td>\n",
       "      <td>-0.006011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_atT/atC</th>\n",
       "      <td>0.023757</td>\n",
       "      <td>0.019454</td>\n",
       "      <td>0.021735</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.005801</td>\n",
       "      <td>0.001213</td>\n",
       "      <td>-0.001445</td>\n",
       "      <td>0.006309</td>\n",
       "      <td>-0.001584</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001629</td>\n",
       "      <td>-0.001510</td>\n",
       "      <td>-0.000686</td>\n",
       "      <td>-0.001157</td>\n",
       "      <td>-0.001090</td>\n",
       "      <td>-0.000755</td>\n",
       "      <td>-0.003092</td>\n",
       "      <td>-0.001008</td>\n",
       "      <td>-0.001417</td>\n",
       "      <td>-0.002926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_cAa/cGa</th>\n",
       "      <td>-0.001481</td>\n",
       "      <td>-0.002130</td>\n",
       "      <td>-0.002480</td>\n",
       "      <td>0.007913</td>\n",
       "      <td>-0.000887</td>\n",
       "      <td>-0.002042</td>\n",
       "      <td>0.004589</td>\n",
       "      <td>-0.000433</td>\n",
       "      <td>-0.003112</td>\n",
       "      <td>-0.000208</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001557</td>\n",
       "      <td>-0.001444</td>\n",
       "      <td>-0.000656</td>\n",
       "      <td>-0.001106</td>\n",
       "      <td>-0.001042</td>\n",
       "      <td>-0.000721</td>\n",
       "      <td>-0.002956</td>\n",
       "      <td>-0.000963</td>\n",
       "      <td>-0.001355</td>\n",
       "      <td>-0.002797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_cCg/cTg</th>\n",
       "      <td>-0.010381</td>\n",
       "      <td>-0.008053</td>\n",
       "      <td>-0.010386</td>\n",
       "      <td>-0.018766</td>\n",
       "      <td>-0.000528</td>\n",
       "      <td>-0.002354</td>\n",
       "      <td>-0.010479</td>\n",
       "      <td>-0.004940</td>\n",
       "      <td>0.004050</td>\n",
       "      <td>-0.004534</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003543</td>\n",
       "      <td>-0.003285</td>\n",
       "      <td>-0.001492</td>\n",
       "      <td>-0.002516</td>\n",
       "      <td>-0.002372</td>\n",
       "      <td>-0.001642</td>\n",
       "      <td>-0.006727</td>\n",
       "      <td>-0.002192</td>\n",
       "      <td>-0.003082</td>\n",
       "      <td>-0.006365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_cGc/cAc</th>\n",
       "      <td>-0.014908</td>\n",
       "      <td>-0.016925</td>\n",
       "      <td>-0.015857</td>\n",
       "      <td>-0.004573</td>\n",
       "      <td>0.013773</td>\n",
       "      <td>0.001683</td>\n",
       "      <td>-0.008026</td>\n",
       "      <td>-0.002648</td>\n",
       "      <td>-0.006579</td>\n",
       "      <td>0.002618</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004430</td>\n",
       "      <td>-0.004107</td>\n",
       "      <td>-0.001866</td>\n",
       "      <td>-0.003146</td>\n",
       "      <td>-0.002966</td>\n",
       "      <td>-0.002053</td>\n",
       "      <td>-0.008410</td>\n",
       "      <td>-0.002741</td>\n",
       "      <td>-0.003854</td>\n",
       "      <td>-0.007958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_cGg/cAg</th>\n",
       "      <td>-0.016014</td>\n",
       "      <td>-0.014547</td>\n",
       "      <td>-0.017854</td>\n",
       "      <td>-0.010275</td>\n",
       "      <td>0.008002</td>\n",
       "      <td>-0.006284</td>\n",
       "      <td>-0.007456</td>\n",
       "      <td>-0.001426</td>\n",
       "      <td>0.008508</td>\n",
       "      <td>-0.002993</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004838</td>\n",
       "      <td>-0.004486</td>\n",
       "      <td>-0.002037</td>\n",
       "      <td>-0.003435</td>\n",
       "      <td>-0.003239</td>\n",
       "      <td>-0.002242</td>\n",
       "      <td>-0.009185</td>\n",
       "      <td>-0.002993</td>\n",
       "      <td>-0.004208</td>\n",
       "      <td>-0.008690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_cGt/cAt</th>\n",
       "      <td>-0.011615</td>\n",
       "      <td>-0.011187</td>\n",
       "      <td>-0.012826</td>\n",
       "      <td>0.024246</td>\n",
       "      <td>0.007312</td>\n",
       "      <td>0.021151</td>\n",
       "      <td>0.012457</td>\n",
       "      <td>-0.007604</td>\n",
       "      <td>-0.001636</td>\n",
       "      <td>-0.003852</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003845</td>\n",
       "      <td>-0.003565</td>\n",
       "      <td>-0.001619</td>\n",
       "      <td>-0.002730</td>\n",
       "      <td>-0.002574</td>\n",
       "      <td>-0.001782</td>\n",
       "      <td>-0.007300</td>\n",
       "      <td>-0.002379</td>\n",
       "      <td>-0.003345</td>\n",
       "      <td>-0.006907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_cTg/cCg</th>\n",
       "      <td>-0.004607</td>\n",
       "      <td>-0.007959</td>\n",
       "      <td>-0.007971</td>\n",
       "      <td>-0.007056</td>\n",
       "      <td>0.000298</td>\n",
       "      <td>-0.005766</td>\n",
       "      <td>-0.005698</td>\n",
       "      <td>-0.002062</td>\n",
       "      <td>-0.007287</td>\n",
       "      <td>0.002962</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001901</td>\n",
       "      <td>-0.001763</td>\n",
       "      <td>-0.000801</td>\n",
       "      <td>-0.001350</td>\n",
       "      <td>-0.001273</td>\n",
       "      <td>-0.000881</td>\n",
       "      <td>-0.003609</td>\n",
       "      <td>-0.001176</td>\n",
       "      <td>-0.001654</td>\n",
       "      <td>-0.003415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_caT/caC</th>\n",
       "      <td>0.009995</td>\n",
       "      <td>0.005038</td>\n",
       "      <td>0.006637</td>\n",
       "      <td>0.004656</td>\n",
       "      <td>-0.002313</td>\n",
       "      <td>-0.000864</td>\n",
       "      <td>0.004396</td>\n",
       "      <td>0.004096</td>\n",
       "      <td>0.000821</td>\n",
       "      <td>0.003394</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001791</td>\n",
       "      <td>-0.001661</td>\n",
       "      <td>-0.000754</td>\n",
       "      <td>-0.001272</td>\n",
       "      <td>-0.001199</td>\n",
       "      <td>-0.000830</td>\n",
       "      <td>-0.003401</td>\n",
       "      <td>-0.001108</td>\n",
       "      <td>-0.001558</td>\n",
       "      <td>-0.003218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_ccA/ccG</th>\n",
       "      <td>0.018762</td>\n",
       "      <td>0.014969</td>\n",
       "      <td>0.014115</td>\n",
       "      <td>-0.001424</td>\n",
       "      <td>-0.005687</td>\n",
       "      <td>0.004527</td>\n",
       "      <td>0.001490</td>\n",
       "      <td>0.001291</td>\n",
       "      <td>0.003986</td>\n",
       "      <td>-0.002854</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.002397</td>\n",
       "      <td>-0.002222</td>\n",
       "      <td>-0.001009</td>\n",
       "      <td>-0.001702</td>\n",
       "      <td>-0.001605</td>\n",
       "      <td>-0.001111</td>\n",
       "      <td>-0.004550</td>\n",
       "      <td>-0.001483</td>\n",
       "      <td>-0.002085</td>\n",
       "      <td>-0.004306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_ccC/cc</th>\n",
       "      <td>-0.003934</td>\n",
       "      <td>-0.003802</td>\n",
       "      <td>-0.004018</td>\n",
       "      <td>0.009353</td>\n",
       "      <td>0.003520</td>\n",
       "      <td>-0.006127</td>\n",
       "      <td>-0.003812</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>-0.003024</td>\n",
       "      <td>0.010492</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000635</td>\n",
       "      <td>-0.000589</td>\n",
       "      <td>-0.000268</td>\n",
       "      <td>-0.000451</td>\n",
       "      <td>-0.000425</td>\n",
       "      <td>-0.000294</td>\n",
       "      <td>-0.001206</td>\n",
       "      <td>-0.000393</td>\n",
       "      <td>-0.000553</td>\n",
       "      <td>-0.001141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_ccG/ccA</th>\n",
       "      <td>0.004290</td>\n",
       "      <td>0.005071</td>\n",
       "      <td>0.003468</td>\n",
       "      <td>-0.015021</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>-0.008363</td>\n",
       "      <td>-0.008740</td>\n",
       "      <td>-0.003240</td>\n",
       "      <td>0.006598</td>\n",
       "      <td>-0.005207</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004047</td>\n",
       "      <td>-0.003752</td>\n",
       "      <td>-0.001704</td>\n",
       "      <td>-0.002874</td>\n",
       "      <td>-0.002709</td>\n",
       "      <td>-0.001875</td>\n",
       "      <td>-0.007683</td>\n",
       "      <td>-0.002504</td>\n",
       "      <td>-0.003520</td>\n",
       "      <td>-0.007270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_ccT/cc</th>\n",
       "      <td>-0.004055</td>\n",
       "      <td>-0.003930</td>\n",
       "      <td>-0.004141</td>\n",
       "      <td>-0.003870</td>\n",
       "      <td>-0.000608</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>-0.003929</td>\n",
       "      <td>-0.003464</td>\n",
       "      <td>0.001989</td>\n",
       "      <td>-0.000955</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000655</td>\n",
       "      <td>-0.000607</td>\n",
       "      <td>-0.000276</td>\n",
       "      <td>-0.000465</td>\n",
       "      <td>-0.000438</td>\n",
       "      <td>-0.000303</td>\n",
       "      <td>-0.001243</td>\n",
       "      <td>-0.000405</td>\n",
       "      <td>-0.000570</td>\n",
       "      <td>-0.001176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_gAc/gCc</th>\n",
       "      <td>-0.003623</td>\n",
       "      <td>-0.003855</td>\n",
       "      <td>-0.003794</td>\n",
       "      <td>-0.004766</td>\n",
       "      <td>0.009058</td>\n",
       "      <td>-0.001557</td>\n",
       "      <td>-0.001820</td>\n",
       "      <td>-0.004446</td>\n",
       "      <td>0.003957</td>\n",
       "      <td>-0.000230</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000841</td>\n",
       "      <td>-0.000779</td>\n",
       "      <td>-0.000354</td>\n",
       "      <td>-0.000597</td>\n",
       "      <td>-0.000563</td>\n",
       "      <td>-0.000389</td>\n",
       "      <td>-0.001596</td>\n",
       "      <td>-0.000520</td>\n",
       "      <td>-0.000731</td>\n",
       "      <td>-0.001510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_gcT/gcC</th>\n",
       "      <td>0.028095</td>\n",
       "      <td>0.033228</td>\n",
       "      <td>0.032076</td>\n",
       "      <td>0.007433</td>\n",
       "      <td>0.001226</td>\n",
       "      <td>0.001435</td>\n",
       "      <td>0.005329</td>\n",
       "      <td>-0.005465</td>\n",
       "      <td>-0.000485</td>\n",
       "      <td>0.005020</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001928</td>\n",
       "      <td>-0.001787</td>\n",
       "      <td>-0.000812</td>\n",
       "      <td>-0.001369</td>\n",
       "      <td>-0.001291</td>\n",
       "      <td>-0.000893</td>\n",
       "      <td>-0.003660</td>\n",
       "      <td>-0.001193</td>\n",
       "      <td>-0.001677</td>\n",
       "      <td>-0.003463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_ggC/ggT</th>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.005507</td>\n",
       "      <td>0.007110</td>\n",
       "      <td>-0.010811</td>\n",
       "      <td>0.006004</td>\n",
       "      <td>-0.006444</td>\n",
       "      <td>-0.009027</td>\n",
       "      <td>0.005564</td>\n",
       "      <td>-0.002184</td>\n",
       "      <td>-0.004605</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003384</td>\n",
       "      <td>-0.003138</td>\n",
       "      <td>-0.001425</td>\n",
       "      <td>-0.002403</td>\n",
       "      <td>-0.002266</td>\n",
       "      <td>-0.001568</td>\n",
       "      <td>-0.006425</td>\n",
       "      <td>-0.002094</td>\n",
       "      <td>-0.002944</td>\n",
       "      <td>-0.006079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_gtA/gtG</th>\n",
       "      <td>0.014077</td>\n",
       "      <td>0.012915</td>\n",
       "      <td>0.010473</td>\n",
       "      <td>0.010295</td>\n",
       "      <td>-0.004484</td>\n",
       "      <td>0.011046</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.001061</td>\n",
       "      <td>0.007070</td>\n",
       "      <td>0.002218</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001597</td>\n",
       "      <td>-0.001481</td>\n",
       "      <td>-0.000673</td>\n",
       "      <td>-0.001134</td>\n",
       "      <td>-0.001069</td>\n",
       "      <td>-0.000740</td>\n",
       "      <td>-0.003032</td>\n",
       "      <td>-0.000988</td>\n",
       "      <td>-0.001389</td>\n",
       "      <td>-0.002869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_tCa/tAa</th>\n",
       "      <td>-0.004816</td>\n",
       "      <td>-0.004669</td>\n",
       "      <td>-0.004918</td>\n",
       "      <td>0.005649</td>\n",
       "      <td>0.001142</td>\n",
       "      <td>0.001927</td>\n",
       "      <td>0.009256</td>\n",
       "      <td>-0.000217</td>\n",
       "      <td>-0.003704</td>\n",
       "      <td>0.003706</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000778</td>\n",
       "      <td>-0.000721</td>\n",
       "      <td>-0.000328</td>\n",
       "      <td>-0.000553</td>\n",
       "      <td>-0.000521</td>\n",
       "      <td>-0.000361</td>\n",
       "      <td>-0.001477</td>\n",
       "      <td>-0.000481</td>\n",
       "      <td>-0.000677</td>\n",
       "      <td>-0.001398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_tCt/tGt</th>\n",
       "      <td>-0.007952</td>\n",
       "      <td>-0.007627</td>\n",
       "      <td>-0.008289</td>\n",
       "      <td>0.002127</td>\n",
       "      <td>-0.004973</td>\n",
       "      <td>0.007609</td>\n",
       "      <td>0.014881</td>\n",
       "      <td>0.000543</td>\n",
       "      <td>-0.007826</td>\n",
       "      <td>0.001511</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.001524</td>\n",
       "      <td>-0.000692</td>\n",
       "      <td>-0.001168</td>\n",
       "      <td>-0.001101</td>\n",
       "      <td>-0.000762</td>\n",
       "      <td>-0.003121</td>\n",
       "      <td>-0.001017</td>\n",
       "      <td>-0.001430</td>\n",
       "      <td>-0.002953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_tGg/tAg</th>\n",
       "      <td>-0.008781</td>\n",
       "      <td>-0.008528</td>\n",
       "      <td>-0.008683</td>\n",
       "      <td>-0.002202</td>\n",
       "      <td>0.002776</td>\n",
       "      <td>-0.003856</td>\n",
       "      <td>-0.002030</td>\n",
       "      <td>-0.002085</td>\n",
       "      <td>-0.005060</td>\n",
       "      <td>0.015828</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001524</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.000642</td>\n",
       "      <td>-0.001082</td>\n",
       "      <td>-0.001021</td>\n",
       "      <td>-0.000706</td>\n",
       "      <td>-0.002894</td>\n",
       "      <td>-0.000943</td>\n",
       "      <td>-0.001326</td>\n",
       "      <td>-0.002738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_tTa/tGa</th>\n",
       "      <td>-0.004284</td>\n",
       "      <td>-0.004155</td>\n",
       "      <td>-0.004378</td>\n",
       "      <td>0.005689</td>\n",
       "      <td>-0.001062</td>\n",
       "      <td>-0.001377</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.000720</td>\n",
       "      <td>-0.003295</td>\n",
       "      <td>0.012290</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000692</td>\n",
       "      <td>-0.000642</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.000492</td>\n",
       "      <td>-0.000464</td>\n",
       "      <td>-0.000321</td>\n",
       "      <td>-0.001314</td>\n",
       "      <td>-0.000428</td>\n",
       "      <td>-0.000602</td>\n",
       "      <td>-0.001244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_taC/taA</th>\n",
       "      <td>-0.007166</td>\n",
       "      <td>-0.006839</td>\n",
       "      <td>-0.007121</td>\n",
       "      <td>-0.000901</td>\n",
       "      <td>-0.005684</td>\n",
       "      <td>-0.001826</td>\n",
       "      <td>-0.002363</td>\n",
       "      <td>-0.003575</td>\n",
       "      <td>-0.002691</td>\n",
       "      <td>0.004036</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001168</td>\n",
       "      <td>-0.001082</td>\n",
       "      <td>-0.000492</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.000782</td>\n",
       "      <td>-0.000541</td>\n",
       "      <td>-0.002216</td>\n",
       "      <td>-0.000722</td>\n",
       "      <td>-0.001016</td>\n",
       "      <td>-0.002097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_taC/taG</th>\n",
       "      <td>-0.006793</td>\n",
       "      <td>-0.006537</td>\n",
       "      <td>-0.006902</td>\n",
       "      <td>0.000648</td>\n",
       "      <td>-0.002868</td>\n",
       "      <td>0.001058</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>-0.003064</td>\n",
       "      <td>-0.002200</td>\n",
       "      <td>0.009554</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001101</td>\n",
       "      <td>-0.001021</td>\n",
       "      <td>-0.000464</td>\n",
       "      <td>-0.000782</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.000510</td>\n",
       "      <td>-0.002090</td>\n",
       "      <td>-0.000681</td>\n",
       "      <td>-0.000957</td>\n",
       "      <td>-0.001977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_taT/taA</th>\n",
       "      <td>-0.004711</td>\n",
       "      <td>-0.004571</td>\n",
       "      <td>-0.004817</td>\n",
       "      <td>0.001360</td>\n",
       "      <td>0.001387</td>\n",
       "      <td>0.004696</td>\n",
       "      <td>-0.001015</td>\n",
       "      <td>-0.004030</td>\n",
       "      <td>0.000764</td>\n",
       "      <td>-0.002210</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000762</td>\n",
       "      <td>-0.000706</td>\n",
       "      <td>-0.000321</td>\n",
       "      <td>-0.000541</td>\n",
       "      <td>-0.000510</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.001446</td>\n",
       "      <td>-0.000471</td>\n",
       "      <td>-0.000663</td>\n",
       "      <td>-0.001368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_tcG/tcA</th>\n",
       "      <td>0.004803</td>\n",
       "      <td>0.005235</td>\n",
       "      <td>0.005580</td>\n",
       "      <td>-0.002478</td>\n",
       "      <td>-0.004161</td>\n",
       "      <td>-0.010003</td>\n",
       "      <td>-0.004765</td>\n",
       "      <td>0.007930</td>\n",
       "      <td>0.003458</td>\n",
       "      <td>-0.005767</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003121</td>\n",
       "      <td>-0.002894</td>\n",
       "      <td>-0.001314</td>\n",
       "      <td>-0.002216</td>\n",
       "      <td>-0.002090</td>\n",
       "      <td>-0.001446</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.001931</td>\n",
       "      <td>-0.002715</td>\n",
       "      <td>-0.005607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_tgC/tgA</th>\n",
       "      <td>-0.006221</td>\n",
       "      <td>-0.006065</td>\n",
       "      <td>-0.006394</td>\n",
       "      <td>-0.010072</td>\n",
       "      <td>-0.004369</td>\n",
       "      <td>-0.002593</td>\n",
       "      <td>-0.003440</td>\n",
       "      <td>-0.005381</td>\n",
       "      <td>-0.001553</td>\n",
       "      <td>0.002221</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001017</td>\n",
       "      <td>-0.000943</td>\n",
       "      <td>-0.000428</td>\n",
       "      <td>-0.000722</td>\n",
       "      <td>-0.000681</td>\n",
       "      <td>-0.000471</td>\n",
       "      <td>-0.001931</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.000885</td>\n",
       "      <td>-0.001827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_tgG/tgA</th>\n",
       "      <td>-0.008826</td>\n",
       "      <td>-0.008504</td>\n",
       "      <td>-0.008986</td>\n",
       "      <td>-0.002022</td>\n",
       "      <td>0.000804</td>\n",
       "      <td>-0.006088</td>\n",
       "      <td>-0.006686</td>\n",
       "      <td>-0.005442</td>\n",
       "      <td>-0.004467</td>\n",
       "      <td>0.011584</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001430</td>\n",
       "      <td>-0.001326</td>\n",
       "      <td>-0.000602</td>\n",
       "      <td>-0.001016</td>\n",
       "      <td>-0.000957</td>\n",
       "      <td>-0.000663</td>\n",
       "      <td>-0.002715</td>\n",
       "      <td>-0.000885</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.002569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codons_ttC/ttT</th>\n",
       "      <td>0.009074</td>\n",
       "      <td>0.005833</td>\n",
       "      <td>0.006200</td>\n",
       "      <td>-0.004718</td>\n",
       "      <td>0.004612</td>\n",
       "      <td>-0.002884</td>\n",
       "      <td>-0.006659</td>\n",
       "      <td>0.007095</td>\n",
       "      <td>-0.000402</td>\n",
       "      <td>-0.001269</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.002953</td>\n",
       "      <td>-0.002738</td>\n",
       "      <td>-0.001244</td>\n",
       "      <td>-0.002097</td>\n",
       "      <td>-0.001977</td>\n",
       "      <td>-0.001368</td>\n",
       "      <td>-0.005607</td>\n",
       "      <td>-0.001827</td>\n",
       "      <td>-0.002569</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1871 rows × 1871 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  AF_ESP   AF_EXAC    AF_TGP   LoFtool   CHROM_1   CHROM_2  \\\n",
       "AF_ESP          1.000000  0.851870  0.807741  0.024530  0.023263 -0.021822   \n",
       "AF_EXAC         0.851870  1.000000  0.805634  0.023114  0.016571 -0.018818   \n",
       "AF_TGP          0.807741  0.805634  1.000000  0.025189  0.018733 -0.023872   \n",
       "LoFtool         0.024530  0.023114  0.025189  1.000000 -0.042282  0.186457   \n",
       "CHROM_1         0.023263  0.016571  0.018733 -0.042282  1.000000 -0.105889   \n",
       "CHROM_2        -0.021822 -0.018818 -0.023872  0.186457 -0.105889  1.000000   \n",
       "CHROM_5         0.006211 -0.001123  0.005495  0.015560 -0.065886 -0.095132   \n",
       "CHROM_7         0.001599  0.003508  0.001747 -0.041412 -0.058084 -0.083867   \n",
       "CHROM_10        0.011395  0.014726  0.019362 -0.009614 -0.052266 -0.075466   \n",
       "CHROM_11       -0.022898 -0.020768 -0.022115  0.121785 -0.076743 -0.110809   \n",
       "CHROM_13       -0.027537 -0.027953 -0.028789 -0.119747 -0.055702 -0.080427   \n",
       "CHROM_14        0.009595  0.005834  0.011515 -0.029261 -0.047855 -0.069097   \n",
       "CHROM_15        0.004683  0.003479  0.007204 -0.024928 -0.044647 -0.064465   \n",
       "CHROM_16       -0.009229 -0.013755 -0.012652  0.042957 -0.057059 -0.082387   \n",
       "CHROM_17       -0.030896 -0.031693 -0.026714 -0.091613 -0.081337 -0.117441   \n",
       "CHROM_20        0.016523  0.022718  0.018930 -0.040403 -0.029393 -0.042440   \n",
       "CHROM_21        0.033424  0.026051  0.027190 -0.033917 -0.030997 -0.044756   \n",
       "CHROM_22        0.003739  0.002169  0.005391 -0.011221 -0.036049 -0.052050   \n",
       "CHROM_MT       -0.003934 -0.003814 -0.004018 -0.007821 -0.004243 -0.006127   \n",
       "REF_A           0.018794  0.016609  0.012548  0.022992 -0.018791  0.009397   \n",
       "REF_AC         -0.007512 -0.005643 -0.005694 -0.004787 -0.007711 -0.004167   \n",
       "REF_C          -0.010447 -0.011472 -0.011046 -0.022567  0.012841 -0.003046   \n",
       "REF_CAAA       -0.001703 -0.001630 -0.001694  0.005030 -0.001837 -0.002653   \n",
       "REF_CAG        -0.005832 -0.005272 -0.005566  0.000437  0.003295 -0.006088   \n",
       "REF_CG         -0.007103 -0.006876 -0.007484 -0.011411 -0.001964 -0.007117   \n",
       "REF_CT         -0.005472 -0.006068 -0.008346  0.001048 -0.006380 -0.008969   \n",
       "REF_CTT        -0.002702 -0.000383 -0.001014  0.008704  0.002694  0.000045   \n",
       "REF_G          -0.012746 -0.010841 -0.010406 -0.032604  0.006412 -0.013088   \n",
       "REF_GA         -0.001044  0.000436  0.003071 -0.001055 -0.003611  0.004763   \n",
       "REF_GC         -0.009958 -0.009661 -0.008387 -0.007193 -0.005438 -0.004334   \n",
       "...                  ...       ...       ...       ...       ...       ...   \n",
       "Codons_agC/agA -0.003750 -0.003521 -0.004138 -0.002935 -0.003778  0.000435   \n",
       "Codons_agT/agC  0.014429  0.013989  0.008679  0.007449  0.003616 -0.003498   \n",
       "Codons_atC/atT  0.000906  0.000755  0.005853 -0.012172  0.005836 -0.000267   \n",
       "Codons_atT/atC  0.023757  0.019454  0.021735  0.004600  0.005801  0.001213   \n",
       "Codons_cAa/cGa -0.001481 -0.002130 -0.002480  0.007913 -0.000887 -0.002042   \n",
       "Codons_cCg/cTg -0.010381 -0.008053 -0.010386 -0.018766 -0.000528 -0.002354   \n",
       "Codons_cGc/cAc -0.014908 -0.016925 -0.015857 -0.004573  0.013773  0.001683   \n",
       "Codons_cGg/cAg -0.016014 -0.014547 -0.017854 -0.010275  0.008002 -0.006284   \n",
       "Codons_cGt/cAt -0.011615 -0.011187 -0.012826  0.024246  0.007312  0.021151   \n",
       "Codons_cTg/cCg -0.004607 -0.007959 -0.007971 -0.007056  0.000298 -0.005766   \n",
       "Codons_caT/caC  0.009995  0.005038  0.006637  0.004656 -0.002313 -0.000864   \n",
       "Codons_ccA/ccG  0.018762  0.014969  0.014115 -0.001424 -0.005687  0.004527   \n",
       "Codons_ccC/cc  -0.003934 -0.003802 -0.004018  0.009353  0.003520 -0.006127   \n",
       "Codons_ccG/ccA  0.004290  0.005071  0.003468 -0.015021  0.000041 -0.008363   \n",
       "Codons_ccT/cc  -0.004055 -0.003930 -0.004141 -0.003870 -0.000608  0.002088   \n",
       "Codons_gAc/gCc -0.003623 -0.003855 -0.003794 -0.004766  0.009058 -0.001557   \n",
       "Codons_gcT/gcC  0.028095  0.033228  0.032076  0.007433  0.001226  0.001435   \n",
       "Codons_ggC/ggT  0.003671  0.005507  0.007110 -0.010811  0.006004 -0.006444   \n",
       "Codons_gtA/gtG  0.014077  0.012915  0.010473  0.010295 -0.004484  0.011046   \n",
       "Codons_tCa/tAa -0.004816 -0.004669 -0.004918  0.005649  0.001142  0.001927   \n",
       "Codons_tCt/tGt -0.007952 -0.007627 -0.008289  0.002127 -0.004973  0.007609   \n",
       "Codons_tGg/tAg -0.008781 -0.008528 -0.008683 -0.002202  0.002776 -0.003856   \n",
       "Codons_tTa/tGa -0.004284 -0.004155 -0.004378  0.005689 -0.001062 -0.001377   \n",
       "Codons_taC/taA -0.007166 -0.006839 -0.007121 -0.000901 -0.005684 -0.001826   \n",
       "Codons_taC/taG -0.006793 -0.006537 -0.006902  0.000648 -0.002868  0.001058   \n",
       "Codons_taT/taA -0.004711 -0.004571 -0.004817  0.001360  0.001387  0.004696   \n",
       "Codons_tcG/tcA  0.004803  0.005235  0.005580 -0.002478 -0.004161 -0.010003   \n",
       "Codons_tgC/tgA -0.006221 -0.006065 -0.006394 -0.010072 -0.004369 -0.002593   \n",
       "Codons_tgG/tgA -0.008826 -0.008504 -0.008986 -0.002022  0.000804 -0.006088   \n",
       "Codons_ttC/ttT  0.009074  0.005833  0.006200 -0.004718  0.004612 -0.002884   \n",
       "\n",
       "                 CHROM_5   CHROM_7  CHROM_10  CHROM_11       ...        \\\n",
       "AF_ESP          0.006211  0.001599  0.011395 -0.022898       ...         \n",
       "AF_EXAC        -0.001123  0.003508  0.014726 -0.020768       ...         \n",
       "AF_TGP          0.005495  0.001747  0.019362 -0.022115       ...         \n",
       "LoFtool         0.015560 -0.041412 -0.009614  0.121785       ...         \n",
       "CHROM_1        -0.065886 -0.058084 -0.052266 -0.076743       ...         \n",
       "CHROM_2        -0.095132 -0.083867 -0.075466 -0.110809       ...         \n",
       "CHROM_5         1.000000 -0.052183 -0.046956 -0.068947       ...         \n",
       "CHROM_7        -0.052183  1.000000 -0.041396 -0.060782       ...         \n",
       "CHROM_10       -0.046956 -0.041396  1.000000 -0.054694       ...         \n",
       "CHROM_11       -0.068947 -0.060782 -0.054694  1.000000       ...         \n",
       "CHROM_13       -0.050043 -0.044117 -0.039698 -0.058290       ...         \n",
       "CHROM_14       -0.042993 -0.037902 -0.034106 -0.050078       ...         \n",
       "CHROM_15       -0.040111 -0.035362 -0.031819 -0.046721       ...         \n",
       "CHROM_16       -0.051263 -0.045192 -0.040665 -0.059710       ...         \n",
       "CHROM_17       -0.073073 -0.064420 -0.057968 -0.085115       ...         \n",
       "CHROM_20       -0.026407 -0.023280 -0.020948 -0.030758       ...         \n",
       "CHROM_21       -0.027848 -0.024550 -0.022091 -0.032437       ...         \n",
       "CHROM_22       -0.032386 -0.028551 -0.025691 -0.037723       ...         \n",
       "CHROM_MT       -0.003812 -0.003361 -0.003024 -0.004440       ...         \n",
       "REF_A           0.047899 -0.001251  0.001485  0.019456       ...         \n",
       "REF_AC         -0.004634 -0.004765 -0.007678 -0.002440       ...         \n",
       "REF_C          -0.018144 -0.007245 -0.001712 -0.011086       ...         \n",
       "REF_CAAA        0.008194  0.009573 -0.001309 -0.001923       ...         \n",
       "REF_CAG         0.000358 -0.000650 -0.005400  0.012986       ...         \n",
       "REF_CG          0.006179  0.001127 -0.000229  0.009196       ...         \n",
       "REF_CT         -0.004625  0.007594 -0.002862  0.010163       ...         \n",
       "REF_CTT        -0.002107 -0.004602 -0.004141  0.015732       ...         \n",
       "REF_G          -0.025718 -0.001668  0.008588 -0.011207       ...         \n",
       "REF_GA         -0.003730 -0.000004 -0.002800 -0.002764       ...         \n",
       "REF_GC          0.001100 -0.001744 -0.004079  0.006466       ...         \n",
       "...                  ...       ...       ...       ...       ...         \n",
       "Codons_agC/agA -0.002876  0.004511 -0.004537  0.000806       ...         \n",
       "Codons_agT/agC  0.012968 -0.000599 -0.001105  0.005607       ...         \n",
       "Codons_atC/atT -0.011117 -0.003099  0.007204 -0.009120       ...         \n",
       "Codons_atT/atC -0.001445  0.006309 -0.001584  0.000284       ...         \n",
       "Codons_cAa/cGa  0.004589 -0.000433 -0.003112 -0.000208       ...         \n",
       "Codons_cCg/cTg -0.010479 -0.004940  0.004050 -0.004534       ...         \n",
       "Codons_cGc/cAc -0.008026 -0.002648 -0.006579  0.002618       ...         \n",
       "Codons_cGg/cAg -0.007456 -0.001426  0.008508 -0.002993       ...         \n",
       "Codons_cGt/cAt  0.012457 -0.007604 -0.001636 -0.003852       ...         \n",
       "Codons_cTg/cCg -0.005698 -0.002062 -0.007287  0.002962       ...         \n",
       "Codons_caT/caC  0.004396  0.004096  0.000821  0.003394       ...         \n",
       "Codons_ccA/ccG  0.001490  0.001291  0.003986 -0.002854       ...         \n",
       "Codons_ccC/cc  -0.003812  0.001415 -0.003024  0.010492       ...         \n",
       "Codons_ccG/ccA -0.008740 -0.003240  0.006598 -0.005207       ...         \n",
       "Codons_ccT/cc  -0.003929 -0.003464  0.001989 -0.000955       ...         \n",
       "Codons_gAc/gCc -0.001820 -0.004446  0.003957 -0.000230       ...         \n",
       "Codons_gcT/gcC  0.005329 -0.005465 -0.000485  0.005020       ...         \n",
       "Codons_ggC/ggT -0.009027  0.005564 -0.002184 -0.004605       ...         \n",
       "Codons_gtA/gtG  0.004000  0.001061  0.007070  0.002218       ...         \n",
       "Codons_tCa/tAa  0.009256 -0.000217 -0.003704  0.003706       ...         \n",
       "Codons_tCt/tGt  0.014881  0.000543 -0.007826  0.001511       ...         \n",
       "Codons_tGg/tAg -0.002030 -0.002085 -0.005060  0.015828       ...         \n",
       "Codons_tTa/tGa  0.003671  0.000720 -0.003295  0.012290       ...         \n",
       "Codons_taC/taA -0.002363 -0.003575 -0.002691  0.004036       ...         \n",
       "Codons_taC/taG  0.000782 -0.003064 -0.002200  0.009554       ...         \n",
       "Codons_taT/taA -0.001015 -0.004030  0.000764 -0.002210       ...         \n",
       "Codons_tcG/tcA -0.004765  0.007930  0.003458 -0.005767       ...         \n",
       "Codons_tgC/tgA -0.003440 -0.005381 -0.001553  0.002221       ...         \n",
       "Codons_tgG/tgA -0.006686 -0.005442 -0.004467  0.011584       ...         \n",
       "Codons_ttC/ttT -0.006659  0.007095 -0.000402 -0.001269       ...         \n",
       "\n",
       "                Codons_tCt/tGt  Codons_tGg/tAg  Codons_tTa/tGa  \\\n",
       "AF_ESP               -0.007952       -0.008781       -0.004284   \n",
       "AF_EXAC              -0.007627       -0.008528       -0.004155   \n",
       "AF_TGP               -0.008289       -0.008683       -0.004378   \n",
       "LoFtool               0.002127       -0.002202        0.005689   \n",
       "CHROM_1              -0.004973        0.002776       -0.001062   \n",
       "CHROM_2               0.007609       -0.003856       -0.001377   \n",
       "CHROM_5               0.014881       -0.002030        0.003671   \n",
       "CHROM_7               0.000543       -0.002085        0.000720   \n",
       "CHROM_10             -0.007826       -0.005060       -0.003295   \n",
       "CHROM_11              0.001511        0.015828        0.012290   \n",
       "CHROM_13              0.010863        0.004692        0.005596   \n",
       "CHROM_14             -0.002742        0.002895       -0.003017   \n",
       "CHROM_15              0.000398       -0.006198        0.002784   \n",
       "CHROM_16              0.000848        0.000181        0.000857   \n",
       "CHROM_17              0.011205       -0.000909        0.001396   \n",
       "CHROM_20             -0.004401       -0.004080       -0.001853   \n",
       "CHROM_21             -0.004641       -0.004303       -0.001954   \n",
       "CHROM_22              0.000397       -0.001880       -0.002273   \n",
       "CHROM_MT             -0.000635       -0.000589       -0.000268   \n",
       "REF_A                -0.017102       -0.015856        0.012876   \n",
       "REF_AC               -0.001613       -0.001496       -0.000679   \n",
       "REF_C                 0.017045        0.016662       -0.012102   \n",
       "REF_CAAA             -0.000275       -0.000255       -0.000116   \n",
       "REF_CAG              -0.001135       -0.001052       -0.000478   \n",
       "REF_CG               -0.001210       -0.001122       -0.000510   \n",
       "REF_CT               -0.001784       -0.001654       -0.000751   \n",
       "REF_CTT              -0.000870       -0.000807       -0.000366   \n",
       "REF_G                 0.012060        0.010319       -0.011921   \n",
       "REF_GA               -0.001516       -0.001406       -0.000638   \n",
       "REF_GC               -0.001690       -0.001567       -0.000712   \n",
       "...                        ...             ...             ...   \n",
       "Codons_agC/agA       -0.000953       -0.000884       -0.000401   \n",
       "Codons_agT/agC       -0.001573       -0.001459       -0.000663   \n",
       "Codons_atC/atT       -0.003346       -0.003103       -0.001409   \n",
       "Codons_atT/atC       -0.001629       -0.001510       -0.000686   \n",
       "Codons_cAa/cGa       -0.001557       -0.001444       -0.000656   \n",
       "Codons_cCg/cTg       -0.003543       -0.003285       -0.001492   \n",
       "Codons_cGc/cAc       -0.004430       -0.004107       -0.001866   \n",
       "Codons_cGg/cAg       -0.004838       -0.004486       -0.002037   \n",
       "Codons_cGt/cAt       -0.003845       -0.003565       -0.001619   \n",
       "Codons_cTg/cCg       -0.001901       -0.001763       -0.000801   \n",
       "Codons_caT/caC       -0.001791       -0.001661       -0.000754   \n",
       "Codons_ccA/ccG       -0.002397       -0.002222       -0.001009   \n",
       "Codons_ccC/cc        -0.000635       -0.000589       -0.000268   \n",
       "Codons_ccG/ccA       -0.004047       -0.003752       -0.001704   \n",
       "Codons_ccT/cc        -0.000655       -0.000607       -0.000276   \n",
       "Codons_gAc/gCc       -0.000841       -0.000779       -0.000354   \n",
       "Codons_gcT/gcC       -0.001928       -0.001787       -0.000812   \n",
       "Codons_ggC/ggT       -0.003384       -0.003138       -0.001425   \n",
       "Codons_gtA/gtG       -0.001597       -0.001481       -0.000673   \n",
       "Codons_tCa/tAa       -0.000778       -0.000721       -0.000328   \n",
       "Codons_tCt/tGt        1.000000       -0.001524       -0.000692   \n",
       "Codons_tGg/tAg       -0.001524        1.000000       -0.000642   \n",
       "Codons_tTa/tGa       -0.000692       -0.000642        1.000000   \n",
       "Codons_taC/taA       -0.001168       -0.001082       -0.000492   \n",
       "Codons_taC/taG       -0.001101       -0.001021       -0.000464   \n",
       "Codons_taT/taA       -0.000762       -0.000706       -0.000321   \n",
       "Codons_tcG/tcA       -0.003121       -0.002894       -0.001314   \n",
       "Codons_tgC/tgA       -0.001017       -0.000943       -0.000428   \n",
       "Codons_tgG/tgA       -0.001430       -0.001326       -0.000602   \n",
       "Codons_ttC/ttT       -0.002953       -0.002738       -0.001244   \n",
       "\n",
       "                Codons_taC/taA  Codons_taC/taG  Codons_taT/taA  \\\n",
       "AF_ESP               -0.007166       -0.006793       -0.004711   \n",
       "AF_EXAC              -0.006839       -0.006537       -0.004571   \n",
       "AF_TGP               -0.007121       -0.006902       -0.004817   \n",
       "LoFtool              -0.000901        0.000648        0.001360   \n",
       "CHROM_1              -0.005684       -0.002868        0.001387   \n",
       "CHROM_2              -0.001826        0.001058        0.004696   \n",
       "CHROM_5              -0.002363        0.000782       -0.001015   \n",
       "CHROM_7              -0.003575       -0.003064       -0.004030   \n",
       "CHROM_10             -0.002691       -0.002200        0.000764   \n",
       "CHROM_11              0.004036        0.009554       -0.002210   \n",
       "CHROM_13             -0.003220       -0.005583        0.000275   \n",
       "CHROM_14              0.004247       -0.004797        0.001447   \n",
       "CHROM_15             -0.004747        0.002572        0.001992   \n",
       "CHROM_16             -0.003424       -0.000113        0.004139   \n",
       "CHROM_17              0.001029        0.004163       -0.005643   \n",
       "CHROM_20             -0.003125        0.002326       -0.002039   \n",
       "CHROM_21             -0.003296       -0.003107       -0.002150   \n",
       "CHROM_22              0.004320       -0.003613       -0.002501   \n",
       "CHROM_MT             -0.000451       -0.000425       -0.000294   \n",
       "REF_A                -0.012144       -0.011449        0.014888   \n",
       "REF_AC               -0.001145       -0.001080       -0.000747   \n",
       "REF_C                 0.006717        0.008330       -0.013316   \n",
       "REF_CAAA             -0.000195       -0.000184       -0.000127   \n",
       "REF_CAG              -0.000806       -0.000760       -0.000526   \n",
       "REF_CG               -0.000859       -0.000810       -0.000561   \n",
       "REF_CT               -0.001267       -0.001195       -0.000827   \n",
       "REF_CTT              -0.000618       -0.000582       -0.000403   \n",
       "REF_G                 0.013979        0.011171       -0.013116   \n",
       "REF_GA               -0.001077       -0.001015       -0.000702   \n",
       "REF_GC               -0.001200       -0.001131       -0.000783   \n",
       "...                        ...             ...             ...   \n",
       "Codons_agC/agA       -0.000677       -0.000638       -0.000442   \n",
       "Codons_agT/agC       -0.001117       -0.001053       -0.000729   \n",
       "Codons_atC/atT       -0.002376       -0.002240       -0.001550   \n",
       "Codons_atT/atC       -0.001157       -0.001090       -0.000755   \n",
       "Codons_cAa/cGa       -0.001106       -0.001042       -0.000721   \n",
       "Codons_cCg/cTg       -0.002516       -0.002372       -0.001642   \n",
       "Codons_cGc/cAc       -0.003146       -0.002966       -0.002053   \n",
       "Codons_cGg/cAg       -0.003435       -0.003239       -0.002242   \n",
       "Codons_cGt/cAt       -0.002730       -0.002574       -0.001782   \n",
       "Codons_cTg/cCg       -0.001350       -0.001273       -0.000881   \n",
       "Codons_caT/caC       -0.001272       -0.001199       -0.000830   \n",
       "Codons_ccA/ccG       -0.001702       -0.001605       -0.001111   \n",
       "Codons_ccC/cc        -0.000451       -0.000425       -0.000294   \n",
       "Codons_ccG/ccA       -0.002874       -0.002709       -0.001875   \n",
       "Codons_ccT/cc        -0.000465       -0.000438       -0.000303   \n",
       "Codons_gAc/gCc       -0.000597       -0.000563       -0.000389   \n",
       "Codons_gcT/gcC       -0.001369       -0.001291       -0.000893   \n",
       "Codons_ggC/ggT       -0.002403       -0.002266       -0.001568   \n",
       "Codons_gtA/gtG       -0.001134       -0.001069       -0.000740   \n",
       "Codons_tCa/tAa       -0.000553       -0.000521       -0.000361   \n",
       "Codons_tCt/tGt       -0.001168       -0.001101       -0.000762   \n",
       "Codons_tGg/tAg       -0.001082       -0.001021       -0.000706   \n",
       "Codons_tTa/tGa       -0.000492       -0.000464       -0.000321   \n",
       "Codons_taC/taA        1.000000       -0.000782       -0.000541   \n",
       "Codons_taC/taG       -0.000782        1.000000       -0.000510   \n",
       "Codons_taT/taA       -0.000541       -0.000510        1.000000   \n",
       "Codons_tcG/tcA       -0.002216       -0.002090       -0.001446   \n",
       "Codons_tgC/tgA       -0.000722       -0.000681       -0.000471   \n",
       "Codons_tgG/tgA       -0.001016       -0.000957       -0.000663   \n",
       "Codons_ttC/ttT       -0.002097       -0.001977       -0.001368   \n",
       "\n",
       "                Codons_tcG/tcA  Codons_tgC/tgA  Codons_tgG/tgA  Codons_ttC/ttT  \n",
       "AF_ESP                0.004803       -0.006221       -0.008826        0.009074  \n",
       "AF_EXAC               0.005235       -0.006065       -0.008504        0.005833  \n",
       "AF_TGP                0.005580       -0.006394       -0.008986        0.006200  \n",
       "LoFtool              -0.002478       -0.010072       -0.002022       -0.004718  \n",
       "CHROM_1              -0.004161       -0.004369        0.000804        0.004612  \n",
       "CHROM_2              -0.010003       -0.002593       -0.006088       -0.002884  \n",
       "CHROM_5              -0.004765       -0.003440       -0.006686       -0.006659  \n",
       "CHROM_7               0.007930       -0.005381       -0.005442        0.007095  \n",
       "CHROM_10              0.003458       -0.001553       -0.004467       -0.000402  \n",
       "CHROM_11             -0.005767        0.002221        0.011584       -0.001269  \n",
       "CHROM_13             -0.014818       -0.002059       -0.000635       -0.013908  \n",
       "CHROM_14              0.015643       -0.004433       -0.006233       -0.007929  \n",
       "CHROM_15              0.001047        0.003488       -0.003102       -0.004093  \n",
       "CHROM_16              0.012596       -0.005286       -0.003115        0.004594  \n",
       "CHROM_17             -0.010023        0.001349        0.005212       -0.001894  \n",
       "CHROM_20              0.000989        0.002981       -0.003828        0.013806  \n",
       "CHROM_21              0.001835       -0.002871       -0.000183       -0.004589  \n",
       "CHROM_22              0.002014        0.001339        0.001963        0.003255  \n",
       "CHROM_MT             -0.001206       -0.000393       -0.000553       -0.001141  \n",
       "REF_A                -0.032467       -0.010581       -0.014877       -0.030720  \n",
       "REF_AC               -0.003062       -0.000998       -0.001403       -0.002897  \n",
       "REF_C                 0.025323        0.013346        0.013766        0.027364  \n",
       "REF_CAAA             -0.000522       -0.000170       -0.000239       -0.000494  \n",
       "REF_CAG              -0.002154       -0.000702       -0.000987       -0.002038  \n",
       "REF_CG               -0.002297       -0.000749       -0.001053       -0.002174  \n",
       "REF_CT               -0.003388       -0.001104       -0.001552       -0.003205  \n",
       "REF_CTT              -0.001652       -0.000538       -0.000757       -0.001563  \n",
       "REF_G                 0.029969        0.004647        0.011558        0.024935  \n",
       "REF_GA               -0.002878       -0.000938       -0.001319       -0.002723  \n",
       "REF_GC               -0.003208       -0.001045       -0.001470       -0.003035  \n",
       "...                        ...             ...             ...             ...  \n",
       "Codons_agC/agA       -0.001809       -0.000590       -0.000829       -0.001712  \n",
       "Codons_agT/agC       -0.002987       -0.000973       -0.001369       -0.002826  \n",
       "Codons_atC/atT       -0.006353       -0.002070       -0.002911       -0.006011  \n",
       "Codons_atT/atC       -0.003092       -0.001008       -0.001417       -0.002926  \n",
       "Codons_cAa/cGa       -0.002956       -0.000963       -0.001355       -0.002797  \n",
       "Codons_cCg/cTg       -0.006727       -0.002192       -0.003082       -0.006365  \n",
       "Codons_cGc/cAc       -0.008410       -0.002741       -0.003854       -0.007958  \n",
       "Codons_cGg/cAg       -0.009185       -0.002993       -0.004208       -0.008690  \n",
       "Codons_cGt/cAt       -0.007300       -0.002379       -0.003345       -0.006907  \n",
       "Codons_cTg/cCg       -0.003609       -0.001176       -0.001654       -0.003415  \n",
       "Codons_caT/caC       -0.003401       -0.001108       -0.001558       -0.003218  \n",
       "Codons_ccA/ccG       -0.004550       -0.001483       -0.002085       -0.004306  \n",
       "Codons_ccC/cc        -0.001206       -0.000393       -0.000553       -0.001141  \n",
       "Codons_ccG/ccA       -0.007683       -0.002504       -0.003520       -0.007270  \n",
       "Codons_ccT/cc        -0.001243       -0.000405       -0.000570       -0.001176  \n",
       "Codons_gAc/gCc       -0.001596       -0.000520       -0.000731       -0.001510  \n",
       "Codons_gcT/gcC       -0.003660       -0.001193       -0.001677       -0.003463  \n",
       "Codons_ggC/ggT       -0.006425       -0.002094       -0.002944       -0.006079  \n",
       "Codons_gtA/gtG       -0.003032       -0.000988       -0.001389       -0.002869  \n",
       "Codons_tCa/tAa       -0.001477       -0.000481       -0.000677       -0.001398  \n",
       "Codons_tCt/tGt       -0.003121       -0.001017       -0.001430       -0.002953  \n",
       "Codons_tGg/tAg       -0.002894       -0.000943       -0.001326       -0.002738  \n",
       "Codons_tTa/tGa       -0.001314       -0.000428       -0.000602       -0.001244  \n",
       "Codons_taC/taA       -0.002216       -0.000722       -0.001016       -0.002097  \n",
       "Codons_taC/taG       -0.002090       -0.000681       -0.000957       -0.001977  \n",
       "Codons_taT/taA       -0.001446       -0.000471       -0.000663       -0.001368  \n",
       "Codons_tcG/tcA        1.000000       -0.001931       -0.002715       -0.005607  \n",
       "Codons_tgC/tgA       -0.001931        1.000000       -0.000885       -0.001827  \n",
       "Codons_tgG/tgA       -0.002715       -0.000885        1.000000       -0.002569  \n",
       "Codons_ttC/ttT       -0.005607       -0.001827       -0.002569        1.000000  \n",
       "\n",
       "[1871 rows x 1871 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "correlation_matrix = X.corr()\n",
    "display(correlation_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of correlated features to drop: 664\n"
     ]
    }
   ],
   "source": [
    "# Correlated features to be dropped\n",
    "upper = correlation_matrix.where(np.triu(np.ones(correlation_matrix.shape), k=1).astype(np.bool))\n",
    "to_drop = [column for column in upper.columns if any(upper[column] > 0.90)]\n",
    "print(f'Number of correlated features to drop: {len(to_drop)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cols = list(X.columns)\n",
    "\n",
    "for col in to_drop: \n",
    "    cols.remove(col)\n",
    "\n",
    "len(cols)\n",
    "\n",
    "X = X[cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we are dropping highly correlated features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 586,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Save the DF for future reference\n",
    "X.to_csv('./data/X_df.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Train / Test splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 571,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Train splits\n",
    "X_train, X_temp, y_train, y_temp = train_test_split(X, Y, test_size=0.40, random_state=42, stratify=Y)\n",
    "\n",
    "# Dev and Test splits\n",
    "X_dev, X_test, y_dev, y_test = train_test_split(X_temp, y_temp, test_size=0.50, random_state=42, stratify=y_temp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are splitting the data into a train, test and dev set."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Logistic Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 572,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average: 0.7687 (+/- 0.0056)\n"
     ]
    }
   ],
   "source": [
    "lr = LogisticRegression()\n",
    "lr_fit = lr.fit(X_train, y_train)\n",
    "\n",
    "# Cross validation\n",
    "scores_log_r = cross_val_score(lr, X_train, y_train, cv=5)\n",
    "\n",
    "print(f'Average: {np.round(np.mean(scores_log_r), 4)} (+/- {np.round(scores_log_r.std()*2, 5)})')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Logistic Regression Grid Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 573,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best Score: 0.7733176518715483\n",
      "Best Parameters: {'C': 21.544346900318832, 'penalty': 'l1'}\n"
     ]
    }
   ],
   "source": [
    "lr_gs = LogisticRegression()\n",
    "\n",
    "# Create regularization penalty space\n",
    "penalty = ['l1', 'l2']\n",
    "\n",
    "# Create regularization hyperparameter space\n",
    "C = np.logspace(0, 4, 10)\n",
    "\n",
    "# Create hyperparameter options\n",
    "hyperparameters = dict(C=C, penalty=penalty)\n",
    "\n",
    "grid_log_r = GridSearchCV(lr_gs, param_grid=hyperparameters, cv=5, n_jobs=-1)\n",
    "grid_log_r.fit(X_train, y_train)\n",
    "\n",
    "print('Best Score:', grid_log_r.best_score_)\n",
    "\n",
    "# Best parameters from the grid search\n",
    "best_params = grid_log_r.best_params_\n",
    "print('Best Parameters:', best_params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 574,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train Accuracy\n",
      "0.7919308652076089\n",
      "Dev Accuracy\n",
      "0.775272281024697\n"
     ]
    }
   ],
   "source": [
    "# Create a model with the best params\n",
    "lr_2 = LogisticRegression(**best_params)\n",
    "\n",
    "# Fit the model\n",
    "lr_2_fit = lr_2.fit(X_train, y_train)\n",
    "\n",
    "print('Train Accuracy')\n",
    "print(lr_2_fit.score(X_train, y_train))\n",
    "\n",
    "print('Dev Accuracy')\n",
    "print(lr_2_fit.score(X_dev, y_dev))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Random Forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 575,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rfc = ensemble.RandomForestClassifier(n_estimators=200, n_jobs=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 578,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cross Val: [0.76722485 0.77566151 0.78061877 0.77588852 0.77103043]\n",
      "Average: 0.7741 (+/- 0.00916)\n"
     ]
    }
   ],
   "source": [
    "# Cross validation\n",
    "rf_acc = cross_val_score(rfc, X_train, y_train, cv=5, n_jobs=-1)\n",
    "\n",
    "print(f'Cross Val: {rf_acc}')\n",
    "print(f'Average: {np.round(np.mean(rf_acc), 4)} (+/- {np.round(rf_acc.std()*2, 5)})')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Random Forest Grid Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 580,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best Score: 0.7479034567396196\n",
      "Best Parameters: {'max_depth': 5, 'max_features': 'sqrt', 'min_samples_split': 2, 'n_estimators': 100}\n"
     ]
    }
   ],
   "source": [
    "parameters = {'n_estimators':[100,300,500],\n",
    "             'max_features':['sqrt', 'log2'],\n",
    "             'min_samples_split':[2,8,20],\n",
    "             'max_depth':[5,6,7,8]}\n",
    "\n",
    "# Instantiating and fitting Grid Search, then printing best score and best parameters\n",
    "grid_rfc = GridSearchCV(rfc, param_grid=parameters, cv=5, n_jobs=2)\n",
    "grid_rfc.fit(X_train, y_train)\n",
    "\n",
    "print('Best Score:', grid_rfc.best_score_)\n",
    "best_params_rfc = grid_rfc.best_params_\n",
    "print('Best Parameters:', best_params_rfc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 581,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train Score: 0.9912\n",
      "Dev Score: 0.7711\n"
     ]
    }
   ],
   "source": [
    "rfc_2 = ensemble.RandomForestClassifier(**best_params_rfc, n_jobs=-1)\n",
    "# Fit the model\n",
    "rfc_2.fit(X_train, y_train)\n",
    "\n",
    "# Train accuracy\n",
    "print(f'Train Score: {np.round(rfc.score(X_train, y_train), 4)}')\n",
    "# Dev accuracy\n",
    "print(f'Dev Score: {np.round(rfc.score(X_dev, y_dev), 4)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Random Forest model and the Logistic Regression model are very similar in their accuracy results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gradient Boosting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 582,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.76671354 0.77540585 0.77000767 0.77358732 0.77435439]\n",
      "Average: 0.772 (+/- 0.00642)\n"
     ]
    }
   ],
   "source": [
    "# We'll make 500 iterations, use 2-deep trees, and set our loss function.\n",
    "params = {'n_estimators': 700,\n",
    "          'max_depth': 3,\n",
    "          'loss': 'deviance'}\n",
    "\n",
    "# Initialize and fit the model.\n",
    "gb_clf = ensemble.GradientBoostingClassifier(**params)\n",
    "gb_acc = cross_val_score(gb_clf, X_train, y_train, cv=5, n_jobs=-1)\n",
    "\n",
    "print(gb_acc)\n",
    "print(f'Average: {np.round(np.mean(gb_acc), 4)} (+/- {np.round(gb_acc.std()*2, 5)})')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gradient Boosting seems to perform slightly better than the Random Forest model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gradient Boosting Grid Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 583,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best Score: 0.7810390672939251\n",
      "Best Parameters: {'loss': 'deviance', 'max_depth': 8, 'max_features': 'sqrt', 'min_samples_split': 10, 'n_estimators': 600}\n"
     ]
    }
   ],
   "source": [
    "parameters = {\n",
    "            'loss':['deviance', 'exponential'],\n",
    "            'min_samples_split':[2, 5, 10],\n",
    "            'max_depth':[5,6,7,8],\n",
    "            'max_features':['sqrt', 'log2'],\n",
    "            'n_estimators':[100,300,600]}\n",
    "\n",
    "#fitting model and printing best parameters and score from model\n",
    "grid_gb = GridSearchCV(gb_clf, param_grid=parameters, n_jobs=2)\n",
    "grid_gb.fit(X_train, y_train)\n",
    "\n",
    "print('Best Score:', grid_gb.best_score_)\n",
    "best_params_gb = grid_gb.best_params_\n",
    "print('Best Parameters:', best_params_gb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 585,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train Score: 0.8279556146451217\n",
      "Dev Score: 0.784706243288848\n",
      "Test Score: 0.7874674029759166\n"
     ]
    }
   ],
   "source": [
    "# We'll make 500 iterations, use 2-deep trees, and set our loss function.\n",
    "params = {'loss': 'deviance', 'max_depth': 8, 'max_features': 'sqrt', 'min_samples_split': 10, 'n_estimators': 600}\n",
    "\n",
    "# Initialize and fit the model.\n",
    "gb_clf_2 = ensemble.GradientBoostingClassifier(**params)\n",
    "gb_clf_2.fit(X_train, y_train)\n",
    "\n",
    "print(f'Train Score: {gb_clf_2.score(X_train, y_train)}')\n",
    "print(f'Dev Score: {gb_clf_2.score(X_dev, y_dev)}')\n",
    "print(f'Test Score: {gb_clf_2.score(X_test, y_test)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gradient Boosting is the most accurate models and seems to be much better in regards generalization."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "\n",
    "While achieving an accuracy level of 78% using Gradient Boosting is not immediately impressive, I think applying machine learning to the field of genomics is a step in the right direction.  While I have no domain knowledge in this specific area, I imagine it would be beneficial to be able to predict with 77% accuracy, wether or not a genetic variation will result in conflicting classifications.  Given this additional knowledge, labs or clinicians could add additional levels of inspection to overcome the increased likelihood of a conflicting classification.\n",
    "\n",
    "## Challenges I faced\n",
    "\n",
    "The most challenging aspect of this problem was preprocessing the data.  \n",
    "\n",
    "* There were quite a few categorical variables that needed to be converted to a binary form.  \n",
    "\n",
    "* Some of the categorical variables also consisted of thousands of unique values.  \n",
    "\n",
    "* Every row consisted of NaN values, which meant imputing the data was a necessity.\n",
    "\n",
    "* I had no domain knowledge in the area, so feature engineering was not an option, which likely hindered my ability to improve the accuracy of my models"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
